<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="tmod">
<title>tmod: Analysis of Gene Set Enrichments • tmod</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="tmod: Analysis of Gene Set Enrichments">
<meta property="og:description" content="tmod">
<meta property="og:image" content="https://january3.github.io/tmod/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">tmod</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.50.12</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/tmod.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/user_manual/tmod_user_manual.html">tmod: A Tool for the Analysis of Gene Set Enrichments</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>tmod: Analysis of Gene Set Enrichments</h1>
                        <h4 data-toc-skip class="author">January Weiner</h4>
            
            <h4 data-toc-skip class="date">2022-09-26</h4>
      
      
      <div class="d-none name"><code>tmod.Rmd</code></div>
    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>The package includes several tools for testing the significance of enrichment of modules or other gene sets as well as visualisation of the features (genes, metabolites etc.) and gene sets, feature sets or transcriptional modules. Furthermore, the package provides blood transcriptional modules described by Chaussabel et al. <span class="citation">(<a href="#ref-chaussabel2008modular" role="doc-biblioref">2008</a>)</span> and by Li et al. <span class="citation">(<a href="#ref-li2014molecular" role="doc-biblioref">2014</a>)</span> as well as metabolic profiling clusters from Weiner et al. <span class="citation">(<a href="#ref-weiner2012biomarkers" role="doc-biblioref">2012</a>)</span>. This user guide is a tutorial and main documentation for the package.</p>
    </div>
    
<p><img src="logo.png" style="position:absolute;top:25px;right:5px;"></p>
<div class="section level2">
<h2 id="before-you-start">Before you start<a class="anchor" aria-label="anchor" href="#before-you-start"></a>
</h2>
<p>A more comprehensive user manual for tmod is found on <a href="https://january3.github.io/tmod/">github</a>. This document is a shortened version of the manual, adapted to ship as a vignette with the package. Therefore, if you are not offline, I strongly suggest to use the online version.</p>
</div>
<div class="section level2">
<h2 id="quick-start-guide">Quick start guide<a class="anchor" aria-label="anchor" href="#quick-start-guide"></a>
</h2>
<p>First, obtain an ordered list of genes. For example, list of genes ordered by q-values (FDR):</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://january3.github.io/tmod/">tmod</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">EgambiaResults</span><span class="op">)</span></span>
<span><span class="va">tt</span> <span class="op">&lt;-</span> <span class="va">EgambiaResults</span></span>
<span><span class="va">tt</span> <span class="op">&lt;-</span> <span class="va">tt</span><span class="op">[</span> <span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">tt</span><span class="op">$</span><span class="va">adj.P.Val</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">l</span>  <span class="op">&lt;-</span> <span class="va">tt</span><span class="op">$</span><span class="va">GENE_SYMBOL</span></span></code></pre></div>
<p>Run the CERNO test:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tmodUtest.html">tmodCERNOtest</a></span><span class="op">(</span><span class="va">l</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                ID                               Title    cerno  N1       AUC</span></span>
<span><span class="co">## LI.M37.0 LI.M37.0 immune activation - generic cluster 426.3578 100 0.7462103</span></span>
<span><span class="co">## DC.M4.2   DC.M4.2                        Inflammation 151.1520  20 0.9503953</span></span>
<span><span class="co">## DC.M3.4   DC.M3.4                          Interferon 129.4727  17 0.8315780</span></span>
<span><span class="co">## DC.M1.2   DC.M1.2                          Interferon 112.7056  17 0.9004196</span></span>
<span><span class="co">## DC.M7.29 DC.M7.29                        Undetermined 118.6759  20 0.8087599</span></span>
<span><span class="co">## LI.M11.0 LI.M11.0          enriched in monocytes (II) 113.8086  20 0.7766542</span></span>
<span><span class="co">##               cES      P.Value    adj.P.Val</span></span>
<span><span class="co">## LI.M37.0 2.131789 1.824844e-18 1.105856e-15</span></span>
<span><span class="co">## DC.M4.2  3.778799 8.040039e-15 2.436132e-12</span></span>
<span><span class="co">## DC.M3.4  3.808019 4.609405e-13 9.310998e-11</span></span>
<span><span class="co">## DC.M1.2  3.314869 2.298170e-10 3.481728e-08</span></span>
<span><span class="co">## DC.M7.29 2.966897 1.002268e-09 1.214749e-07</span></span>
<span><span class="co">## LI.M11.0 2.845216 5.255069e-09 5.307620e-07</span></span></code></pre>
<p>Visualize the results:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ggPanelplot.html">ggPanelplot</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<p><img src="tmod_files/figure-html/qsg3-1.png" width="480"></p>
<p>Inspect individual gene sets using evidence plots:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ggEvidencePlot.html">ggEvidencePlot</a></span><span class="op">(</span><span class="va">l</span>, <span class="st">"LI.M37.0"</span>, gene.labels<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/ggEvidencePlot.html">ggEvidencePlot</a></span><span class="op">(</span><span class="va">l</span>, <span class="st">"DC.M4.2"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/ggEvidencePlot.html">ggEvidencePlot</a></span><span class="op">(</span><span class="va">l</span>, <span class="st">"DC.M3.4"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/ggEvidencePlot.html">ggEvidencePlot</a></span><span class="op">(</span><span class="va">l</span>, <span class="st">"DC.M1.2"</span><span class="op">)</span></span></code></pre></div>
<p><img src="tmod_files/figure-html/unnamed-chunk-2-1.png" width="960"></p>
<p>Find out which genes are in DC.M1.2:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getModuleMembers.html">getModuleMembers</a></span><span class="op">(</span><span class="st">"DC.M1.2"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $DC.M1.2</span></span>
<span><span class="co">##  [1] "LY6E"      "IFIT1"     "OAS1"      "IFIT3"     "OAS3"      "OASL"     </span></span>
<span><span class="co">##  [7] "LOC129607" "ISG15"     "HERC5"     "MX1"       "BATF2"     "LAMP3"    </span></span>
<span><span class="co">## [13] "IFI44L"    "XAF1"      "IFI44"     "OAS2"      "TRIM6"     "HES4"     </span></span>
<span><span class="co">## [19] "OTOF"      "FLJ20035"  "IFITM3"    "CXCL10"    "EPSTI1"    "SERPING1" </span></span>
<span><span class="co">## [25] "LOC26010"  "RSAD2"     "RTP4"</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Gene set enrichment (GSE) analysis is an increasingly important tool in the biological interpretation of high throughput data, versatile and powerful. In general, there are three generations of GSE algorithms and packages.</p>
<p>First generation approaches test for enrichment in defined sets of differentially expressed genes (often called “foreground”) against the set of all genes (“background”). The statistical test involved is usually a hypergeometric or Fisher’s exact test. The main problem with this kind of approach is that it relies on arbitrary thresholds (like p-value or log fold change cut-offs), and the number of genes that go into the “foreground” set depends on the statistical power involved. Comparison between the same experimental condition will thus yield vastly different results depending on the number of samples used in the experiment.</p>
<p>The second generation of GSE involve tests which do not rely on such arbitrary definitions of what is differentially expressed, and what not, and instead directly or indirectly employ the information about the statistical distribution of individual genes. A popular implementation of this type of GSE analysis is the eponymous GSEA program <span class="citation">(<a href="#ref-subramanian2005gene" role="doc-biblioref">Subramanian et al. 2005</a>)</span>. While popular and quite powerful for a range of applications, this software has important limitations due to its reliance on bootstrapping to obtain an exact p-value. For one thing, the performance of GSEA dramatically decreases for small sample numbers <span class="citation">(<a href="#ref-weiner2016tmod" role="doc-biblioref">Weiner 3rd and Domaszewska 2016</a>)</span>. Moreover, the specifics of the approach prevent it from being used in applications where a direct test for differential expression is either not present (for example, in multivariate functional analysis, see Section <a href="#functional-multivariate-analysis">“Functional multivariate analysis”</a>).</p>
<p>The <code>tmod</code> package <span class="citation">(<a href="#ref-zyla2019gene" role="doc-biblioref">Zyla et al. 2019</a>)</span> and the included CERNO<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;Coincident Extreme Ranks in Numerical Observations &lt;span class="citation"&gt;(&lt;a href="#ref-yamaguchi2008ifn" role="doc-biblioref"&gt;Yamaguchi et al. 2008&lt;/a&gt;)&lt;/span&gt;&lt;/p&gt;'><sup>1</sup></a> test belong to the second generation of algorithms. However, unlike the program GSEA, the CERNO relies exclusively on an ordered list of genes, and the test statistic has a χ² distribution. Thus, it is suitable for any application in which an ordered list of genes is generated: for example, it is possible to apply tmod to weights of PCA components or to variable importance measure of a machine learning model.</p>
<p><code>tmod</code> was created with the following properties in mind: (i) test for enrichment which relies on a list of sorted genes, (ii) with an analytical solution, (iii) flexible, allowing custom gene sets and analyses, (iv) with visualizations of multiple analysis results, suitable for time series and suchlike, (v) including transcriptional module definitions not present in other databases and, finally, (vi) to be suitable for use in R.</p>
</div>
<div class="section level2">
<h2 id="dive-into-tmod-analysis-of-transcriptomic-responses-to-tuberculosis">Dive into tmod: analysis of transcriptomic responses to tuberculosis<a class="anchor" aria-label="anchor" href="#dive-into-tmod-analysis-of-transcriptomic-responses-to-tuberculosis"></a>
</h2>
<div class="section level3">
<h3 id="introduction-1">Introduction<a class="anchor" aria-label="anchor" href="#introduction-1"></a>
</h3>
<p>In this chapter, I will use an example data set included in tmod to show the application of tmod to the analysis of differential gene expression. The data set has been generated by Maertzdorf et al. <span class="citation">(<a href="#ref-maertzdorf2011functional" role="doc-biblioref">2011</a>)</span> and has the GEO ID GSE28623. Is based on whole blood RNA microarrays from tuberculosis (TB) patients and healthy controls.</p>
<p>Although microarrays were used to generate the data, the principle is the same as in RNASeq.</p>
</div>
<div class="section level3">
<h3 id="the-gambia-data-set">The Gambia data set<a class="anchor" aria-label="anchor" href="#the-gambia-data-set"></a>
</h3>
<p>In the following, we will use the Egambia data set included in the package. The data is already background corrected and normalized, so we can proceed with a differential gene expression analysis. Note that only a bit over 5000 genes from the original set of over 45000 probes is included.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://january3.github.io/tmod/">tmod</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">Egambia</span><span class="op">)</span></span>
<span><span class="va">E</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">Egambia</span><span class="op">[</span>,<span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>The results for this data set have been obtained using the limma package as follows:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioinf.wehi.edu.au/limma" class="external-link">limma</a></span><span class="op">)</span></span>
<span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span>Intercept<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">30</span><span class="op">)</span>, TB<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>, each<span class="op">=</span> <span class="fl">15</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/ebayes.html" class="external-link">eBayes</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/limma/man/lmFit.html" class="external-link">lmFit</a></span><span class="op">(</span><span class="va">E</span>, <span class="va">design</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">tt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/toptable.html" class="external-link">topTable</a></span><span class="op">(</span><span class="va">fit</span>, coef<span class="op">=</span><span class="fl">2</span>, number<span class="op">=</span><span class="cn">Inf</span>, </span>
<span>  genelist<span class="op">=</span><span class="va">Egambia</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>However, they are also included in the tmod package:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">EgambiaResults</span><span class="op">)</span></span>
<span><span class="va">tt</span> <span class="op">&lt;-</span> <span class="va">EgambiaResults</span></span></code></pre></div>
<p>The table below shows first couple of results from the table <code>tt</code>.</p>
<table style="width:93%;" class="table">
<colgroup>
<col width="19%">
<col width="45%">
<col width="11%">
<col width="16%">
</colgroup>
<thead><tr class="header">
<th align="left">GENE_SYMBOL</th>
<th align="left">GENE_NAME</th>
<th align="right">logFC</th>
<th align="right">adj.P.Val</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">FAM20A</td>
<td align="left">family with sequence similarity 20, member A"</td>
<td align="right">2.956</td>
<td align="right">0.001899</td>
</tr>
<tr class="even">
<td align="left">FCGR1B</td>
<td align="left">Fc fragment of IgG, high affinity Ib, receptor (CD64)"</td>
<td align="right">2.391</td>
<td align="right">0.002095</td>
</tr>
<tr class="odd">
<td align="left">BATF2</td>
<td align="left">basic leucine zipper transcription factor, ATF-like 2</td>
<td align="right">2.681</td>
<td align="right">0.002216</td>
</tr>
<tr class="even">
<td align="left">ANKRD22</td>
<td align="left">ankyrin repeat domain 22</td>
<td align="right">2.764</td>
<td align="right">0.002692</td>
</tr>
<tr class="odd">
<td align="left">SEPT4</td>
<td align="left">septin 4</td>
<td align="right">3.287</td>
<td align="right">0.002692</td>
</tr>
<tr class="even">
<td align="left">CD274</td>
<td align="left">CD274 molecule</td>
<td align="right">2.377</td>
<td align="right">0.002692</td>
</tr>
</tbody>
</table>
<p>OK, we see some of the genes known to be prominent in the human host response to TB. We can display one of these using tmod’s <code>showGene</code> function (it’s just a <code>boxplot</code> combined with a <code>beeswarm</code>, nothing special):</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CTRL"</span>, <span class="st">"TB"</span><span class="op">)</span>, each<span class="op">=</span><span class="fl">15</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/showGene.html">showGene</a></span><span class="op">(</span><span class="va">E</span><span class="op">[</span><span class="st">"20799"</span>,<span class="op">]</span>, <span class="va">group</span>,</span>
<span>  main<span class="op">=</span><span class="va">Egambia</span><span class="op">[</span><span class="st">"20799"</span>, <span class="st">"GENE_SYMBOL"</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="tmod_files/figure-html/basic2-1.png" width="384"></p>
<p>Fine, but what about the gene sets?</p>
</div>
<div class="section level3">
<h3 id="transcriptional-module-analysis-with-gse">Transcriptional module analysis with GSE<a class="anchor" aria-label="anchor" href="#transcriptional-module-analysis-with-gse"></a>
</h3>
<p>There are two main functions in tmod to understand which modules or gene sets are significantly enriched. There are several statistical tests which can be used from within tmod (see chapter <a href="#statistical-tests-in-tmod">“Statistical tests in tmod”</a> below), but here we will use the CERNO test, which is the main reason this package exist. CERNO is particularly fast and robust second generation approach, recommended for most applications.</p>
<p>CERNO works with an ordered list of genes (only ranks matter, no other statistic is necessary); the idea is to test, for each gene set, whether the genes in this gene set are more likely than others to be at the beginning of that list. The CERNO statistic has a <span class="math inline">\(\chi^2\)</span> distribution and therefore no randomization is necessary, making the test really fast.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">l</span>    <span class="op">&lt;-</span> <span class="va">tt</span><span class="op">$</span><span class="va">GENE_SYMBOL</span></span>
<span><span class="va">resC</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tmodUtest.html">tmodCERNOtest</a></span><span class="op">(</span><span class="va">l</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">resC</span>, <span class="fl">15</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                  ID                                      Title cerno  N1   AUC</span></span>
<span><span class="co">## LI.M37.0   LI.M37.0        immune activation - generic cluster 426.4 100 0.746</span></span>
<span><span class="co">## DC.M4.2     DC.M4.2                               Inflammation 151.2  20 0.950</span></span>
<span><span class="co">## DC.M3.4     DC.M3.4                                 Interferon 129.5  17 0.832</span></span>
<span><span class="co">## DC.M1.2     DC.M1.2                                 Interferon 112.7  17 0.900</span></span>
<span><span class="co">## DC.M7.29   DC.M7.29                               Undetermined 118.7  20 0.809</span></span>
<span><span class="co">## LI.M11.0   LI.M11.0                 enriched in monocytes (II) 113.8  20 0.777</span></span>
<span><span class="co">## DC.M3.2     DC.M3.2                               Inflammation 124.0  24 0.836</span></span>
<span><span class="co">## LI.S4         LI.S4                 Monocyte surface signature  76.4  10 0.897</span></span>
<span><span class="co">## LI.M112.0 LI.M112.0                  complement activation (I)  73.7  11 0.846</span></span>
<span><span class="co">## DC.M7.35   DC.M7.35                               Undetermined  82.4  14 0.803</span></span>
<span><span class="co">## DC.M7.16   DC.M7.16                               Undetermined  66.2  10 0.822</span></span>
<span><span class="co">## LI.M75       LI.M75                    antiviral IFN signature  65.3  10 0.893</span></span>
<span><span class="co">## LI.M16       LI.M16             TLR and inflammatory signaling  46.3   5 0.979</span></span>
<span><span class="co">## LI.M67       LI.M67                  activated dendritic cells  49.5   6 0.971</span></span>
<span><span class="co">## LI.M165     LI.M165 enriched in activated dendritic cells (II)  91.7  19 0.720</span></span>
<span><span class="co">##            cES  P.Value adj.P.Val</span></span>
<span><span class="co">## LI.M37.0  2.13 1.82e-18  1.11e-15</span></span>
<span><span class="co">## DC.M4.2   3.78 8.04e-15  2.44e-12</span></span>
<span><span class="co">## DC.M3.4   3.81 4.61e-13  9.31e-11</span></span>
<span><span class="co">## DC.M1.2   3.31 2.30e-10  3.48e-08</span></span>
<span><span class="co">## DC.M7.29  2.97 1.00e-09  1.21e-07</span></span>
<span><span class="co">## LI.M11.0  2.85 5.26e-09  5.31e-07</span></span>
<span><span class="co">## DC.M3.2   2.58 1.21e-08  1.05e-06</span></span>
<span><span class="co">## LI.S4     3.82 1.61e-08  1.22e-06</span></span>
<span><span class="co">## LI.M112.0 3.35 1.72e-07  1.16e-05</span></span>
<span><span class="co">## DC.M7.35  2.94 2.93e-07  1.77e-05</span></span>
<span><span class="co">## DC.M7.16  3.31 7.38e-07  4.06e-05</span></span>
<span><span class="co">## LI.M75    3.26 1.05e-06  5.28e-05</span></span>
<span><span class="co">## LI.M16    4.63 1.25e-06  5.81e-05</span></span>
<span><span class="co">## LI.M67    4.13 1.69e-06  7.32e-05</span></span>
<span><span class="co">## LI.M165   2.41 2.44e-06  9.86e-05</span></span></code></pre>
<p>Only first 15 results are shown above.</p>
<p>Columns in the above table contain the following:</p>
<ul>
<li>
<strong>ID</strong> The module ID. IDs starting with “LI” come from Li et al. <span class="citation">(<a href="#ref-li2014molecular" role="doc-biblioref">Li et al. 2014</a>)</span>, while IDs starting with “DC” have been defined by Chaussabel et al. <span class="citation">(<a href="#ref-chaussabel2008modular" role="doc-biblioref">Chaussabel et al. 2008</a>)</span>.</li>
<li>
<strong>Title</strong> The module description</li>
<li>
<strong>cerno</strong> The CERNO statistic</li>
<li>
<strong>N1</strong> Number of genes in the module</li>
<li>
<strong>AUC</strong> The area under curve – main size estimate</li>
<li>
<strong>cES</strong> CERNO statistic divided by <span class="math inline">\(2\times N1\)</span>
</li>
<li>
<strong>P.Value</strong> P-value from the hypergeometric test</li>
<li>
<strong>adj.P.Val</strong> P-value adjusted for multiple testing using the Benjamini-Hochberg correction</li>
</ul>
<p>These results make a lot of sense: the transcriptional modules found to be enriched in a comparison of TB patients with healthy individuals are in line with the published findings. In especially, we see the interferon response, complement system as well as T-cell related modules.</p>
</div>
<div class="section level3">
<h3 id="visualizing-results">Visualizing results<a class="anchor" aria-label="anchor" href="#visualizing-results"></a>
</h3>
<p>The main working horse for visualizing the results in tmod is the function <code>ggPanelplot</code>. This is really a glorified heatmap which shows both the effect size (size of the blob on the figure below) and the p-value (intensity of the color). Each column corresponds to a different comparison. Here, there will be only one column for the only comparison we made, however we need to wrap it in a <code>list</code> object. However, we can also use a slightly different representation to also show how many significantly up- and down-regulated<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;Formally, we don’t test for regulation here. “Differentially expressed” is the correct expression. I will use, however, the word “regulated” throughout this manual with the understanding that it only means “there is a difference between two conditions” because it is shorter.&lt;/p&gt;"><sup>2</sup></a> genes are found in the enriched modules (right panel on the figure below).</p>
<p>Note: below I use the <code>plot_grid</code> function from the cowplot package to put the figures side by side.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">g1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ggPanelplot.html">ggPanelplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Gambia<span class="op">=</span><span class="va">resC</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## calculate the number of significant genes</span></span>
<span><span class="co">## per module</span></span>
<span><span class="va">sgenes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tmodDecideTests.html">tmodDecideTests</a></span><span class="op">(</span>g<span class="op">=</span><span class="va">tt</span><span class="op">$</span><span class="va">GENE_SYMBOL</span>,</span>
<span>  lfc<span class="op">=</span><span class="va">tt</span><span class="op">$</span><span class="va">logFC</span>,</span>
<span>  pval<span class="op">=</span><span class="va">tt</span><span class="op">$</span><span class="va">adj.P.Val</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">sgenes</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"Gambia"</span></span>
<span><span class="va">g2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ggPanelplot.html">ggPanelplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Gambia<span class="op">=</span><span class="va">resC</span><span class="op">)</span>, sgenes <span class="op">=</span> <span class="va">sgenes</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">g1</span>, <span class="va">g2</span>, labels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="tmod_files/figure-html/panelplot00-1.png" width="1344"></p>
<p>On the right hand side, the red color on the bars indicates that all signficantly regulated in the enriched modules. The size of the bar corresponds to the AUC, and intensity of the color corresponds to the p-value. See chapter “Visualisation and presentation of results in tmod” for more information on this and other functions.</p>
</div>
</div>
<div class="section level2">
<h2 id="statistical-tests-in-tmod">Statistical tests in tmod<a class="anchor" aria-label="anchor" href="#statistical-tests-in-tmod"></a>
</h2>
<div class="section level3">
<h3 id="introduction-2">Introduction<a class="anchor" aria-label="anchor" href="#introduction-2"></a>
</h3>
<p>There is a substantial numer of different gene set enrichment tests. Several are implemented in <code>tmod</code> (see Table below for a summary). This chapter gives an overview of the possibilities for gene set enrichment analysis with <code>tmod</code>.</p>
<table style="width:96%;" class="table">
<colgroup>
<col width="25%">
<col width="31%">
<col width="38%">
</colgroup>
<thead><tr class="header">
<th align="left">Test</th>
<th align="left">Description</th>
<th align="left">Input type</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">tmodHGtest</td>
<td align="left">First generation test</td>
<td align="left">Two sets, foreground and background</td>
</tr>
<tr class="even">
<td align="left">tmodUtest</td>
<td align="left">Wilcoxon U test</td>
<td align="left">Ordered gene list</td>
</tr>
<tr class="odd">
<td align="left">tmodCERNOtest</td>
<td align="left">CERNO test</td>
<td align="left">Ordered gene list</td>
</tr>
<tr class="even">
<td align="left">tmodZtest</td>
<td align="left">variant of the CERNO test</td>
<td align="left">Ordered gene list</td>
</tr>
<tr class="odd">
<td align="left">tmodPLAGEtest</td>
<td align="left">eigengene-based</td>
<td align="left">Expression matrix</td>
</tr>
<tr class="even">
<td align="left">tmodAUC</td>
<td align="left">general permutation based testing</td>
<td align="left">Matrix of ranks</td>
</tr>
<tr class="odd">
<td align="left">tmodGeneSetTest</td>
<td align="left">permutation based on a particular statistic</td>
<td align="left">A statistic (e.g. logFC)</td>
</tr>
</tbody>
</table>
<p>In the following, I will briefly describe the various tests and show examples of usage on the Gambia data set.</p>
</div>
<div class="section level3">
<h3 id="first-generation-tests">First generation tests<a class="anchor" aria-label="anchor" href="#first-generation-tests"></a>
</h3>
<p>First generation tests were based on an overrepresentation analysis (ORA). In essence, they rely on splitting the genes into two groups: the genes of interest (foreground), such as genes that we consider to be significantly regulated in an experimental condition, and all the rest (background). For a given gene set, this results in a <span class="math inline">\(2\times 2\)</span> contingency table. If these two factors are independent (i.e., the probability of a gene belonging to a gene set is independent of the probability of a gene being regulated in the experimental condition), then we can easily derive expected frequencies for each cell of the table. Several statistical tests exist to test whether the expected frequencies differ significantly from the observed frequencies.</p>
<p>In tmod, the function <code><a href="../reference/tmodUtest.html">tmodHGtest()</a></code>, performs a hypergeometric test on two groups of genes: ‘foreground’ (fg), or the list of differentially expressed genes, and ‘background’ (bg) – the gene universe, i.e., all genes present in the analysis. The gene identifiers used currently by tmod are HGNC identifiers, and we will use the GENE_SYMBOL field from the Egambia data set.</p>
<p>In this particular example, however, we have almost no genes which are significantly differentially expressed after correction for multiple testing: the power of the test with 10 individuals in each group is too low. For the sake of the example, we will therefore relax our selection. Normally, I’d use a q-value threshold of at least 0.001.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fg</span> <span class="op">&lt;-</span> <span class="va">tt</span><span class="op">$</span><span class="va">GENE_SYMBOL</span><span class="op">[</span><span class="va">tt</span><span class="op">$</span><span class="va">adj.P.Val</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span> <span class="va">tt</span><span class="op">$</span><span class="va">logFC</span> <span class="op">)</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">resHG</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tmodUtest.html">tmodHGtest</a></span><span class="op">(</span>fg<span class="op">=</span><span class="va">fg</span>, bg<span class="op">=</span><span class="va">tt</span><span class="op">$</span><span class="va">GENE_SYMBOL</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>width<span class="op">=</span><span class="fl">60</span><span class="op">)</span></span>
<span><span class="va">resHG</span></span></code></pre></div>
<pre><code><span><span class="co">##                  ID</span></span>
<span><span class="co">## DC.M3.4     DC.M3.4</span></span>
<span><span class="co">## DC.M4.2     DC.M4.2</span></span>
<span><span class="co">## DC.M5.12   DC.M5.12</span></span>
<span><span class="co">## LI.M112.0 LI.M112.0</span></span>
<span><span class="co">## LI.M11.0   LI.M11.0</span></span>
<span><span class="co">## DC.M7.29   DC.M7.29</span></span>
<span><span class="co">## LI.M75       LI.M75</span></span>
<span><span class="co">## LI.S4         LI.S4</span></span>
<span><span class="co">## DC.M7.35   DC.M7.35</span></span>
<span><span class="co">## LI.S5         LI.S5</span></span>
<span><span class="co">## DC.M1.2     DC.M1.2</span></span>
<span><span class="co">## LI.M165     LI.M165</span></span>
<span><span class="co">## LI.M4.3     LI.M4.3</span></span>
<span><span class="co">## LI.M16       LI.M16</span></span>
<span><span class="co">##                                                      Title</span></span>
<span><span class="co">## DC.M3.4                                         Interferon</span></span>
<span><span class="co">## DC.M4.2                                       Inflammation</span></span>
<span><span class="co">## DC.M5.12                                        Interferon</span></span>
<span><span class="co">## LI.M112.0                        complement activation (I)</span></span>
<span><span class="co">## LI.M11.0                        enriched in monocytes (II)</span></span>
<span><span class="co">## DC.M7.29                                      Undetermined</span></span>
<span><span class="co">## LI.M75                             antiviral IFN signature</span></span>
<span><span class="co">## LI.S4                           Monocyte surface signature</span></span>
<span><span class="co">## DC.M7.35                                      Undetermined</span></span>
<span><span class="co">## LI.S5                                 DC surface signature</span></span>
<span><span class="co">## DC.M1.2                                         Interferon</span></span>
<span><span class="co">## LI.M165         enriched in activated dendritic cells (II)</span></span>
<span><span class="co">## LI.M4.3   myeloid cell enriched receptors and transporters</span></span>
<span><span class="co">## LI.M16                      TLR and inflammatory signaling</span></span>
<span><span class="co">##           b  B  n    N    E  P.Value adj.P.Val</span></span>
<span><span class="co">## DC.M3.4   7 17 47 4826 42.3 9.44e-11  5.72e-08</span></span>
<span><span class="co">## DC.M4.2   6 20 47 4826 30.8 2.15e-08  6.51e-06</span></span>
<span><span class="co">## DC.M5.12  4  7 47 4826 58.7 2.71e-07  5.47e-05</span></span>
<span><span class="co">## LI.M112.0 4 11 47 4826 37.3 2.48e-06  3.76e-04</span></span>
<span><span class="co">## LI.M11.0  4 20 47 4826 20.5 3.41e-05  3.45e-03</span></span>
<span><span class="co">## DC.M7.29  4 20 47 4826 20.5 3.41e-05  3.45e-03</span></span>
<span><span class="co">## LI.M75    3 10 47 4826 30.8 9.91e-05  7.50e-03</span></span>
<span><span class="co">## LI.S4     3 10 47 4826 30.8 9.91e-05  7.50e-03</span></span>
<span><span class="co">## DC.M7.35  3 14 47 4826 22.0 2.92e-04  1.79e-02</span></span>
<span><span class="co">## LI.S5     4 34 47 4826 12.1 2.96e-04  1.79e-02</span></span>
<span><span class="co">## DC.M1.2   3 17 47 4826 18.1 5.35e-04  2.95e-02</span></span>
<span><span class="co">## LI.M165   3 19 47 4826 16.2 7.52e-04  3.80e-02</span></span>
<span><span class="co">## LI.M4.3   2  5 47 4826 41.1 9.11e-04  3.94e-02</span></span>
<span><span class="co">## LI.M16    2  5 47 4826 41.1 9.11e-04  3.94e-02</span></span></code></pre>
<p>The columns in the above table contain the following:</p>
<ul>
<li>
<strong>ID</strong> The module ID. IDs starting with “LI” come from Li et al. <span class="citation">(<a href="#ref-li2014molecular" role="doc-biblioref">Li et al. 2014</a>)</span>, while IDs starting with “DC” have been defined by Chaussabel et al. <span class="citation">(<a href="#ref-chaussabel2008modular" role="doc-biblioref">Chaussabel et al. 2008</a>)</span>.</li>
<li>
<strong>Title</strong> The module description</li>
<li>
<strong>b</strong> Number of genes from the given module in the fg set</li>
<li>
<strong>B</strong> Number of genes from the module in the bg set</li>
<li>
<strong>n</strong> Size of the fg set</li>
<li>
<strong>N</strong> Size of the bg set</li>
<li>
<strong>E</strong> Enrichment, calcualted as (b/n)/(B/N)</li>
<li>
<strong>P.Value</strong> P-value from the hypergeometric test</li>
<li>
<strong>adj.P.Val</strong> P-value adjusted for multiple testing using the Benjamini-Hochberg correction</li>
</ul>
<p>Well, IFN signature in TB is well known. However, the numbers of genes are not high: n is the size of the foreground, and b the number of genes in fg that belong to the given module. N and B are the respective totals – size of bg+fg and number of genes that belong to the module that are found in this totality of the analysed genes. If we were using the full Gambia data set (with all its genes), we would have a different situation.</p>
<p>Lack of significant genes is the main problem of ORA: splitting the genes into foreground and background relies on an arbitrary threshold which will yield very different results for different sample sizes.</p>
</div>
<div class="section level3">
<h3 id="second-generation-tests">Second generation tests<a class="anchor" aria-label="anchor" href="#second-generation-tests"></a>
</h3>
<div class="section level4">
<h4 id="u-test-tmodutest">U-test (<code>tmodUtest</code>)<a class="anchor" aria-label="anchor" href="#u-test-tmodutest"></a>
</h4>
<p>Another approach is to sort all the genes (for example, by the respective p-value) and perform a U-test on the ranks of (i) genes belonging to the module and (ii) genes that do not belong to the module. This is a bit slower, but often works even in the case if the power of the statistical test for differential expression is low. That is, even if only a few genes or none at all are significant at acceptable thresholds, sorting them by the p-value or another similar metric can nonetheless allow to get meaningful enrichments<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;The rationale is that the non-significant p-values are not associated with the test that we are actually performing, but merely used to sort the gene list. Thus, it does not matter whether they are significant or not.&lt;/p&gt;"><sup>3</sup></a>.</p>
<p>Moreover, we do not need to set arbitrary thresholds, like p-value or logFC cutoff.</p>
<p>The main issue with the U-test is that it detects enrichments as well as depletions – that is, modules which are enriched at the bottom of the list (e.g. modules which are never, ever regulated in a particular comparison) will be detected as well. This is often undesirable. Secondly, large modules will be reported as significant even if the actual effect size (i.e., AUC) is modest or very small, just because of the sheer number of genes in a module. Unfortunately, also the reverse is true: modules with a small number of genes, even if they consist of highly up- or down-regulated genes from the top of the list will not be detected.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">l</span>    <span class="op">&lt;-</span> <span class="va">tt</span><span class="op">$</span><span class="va">GENE_SYMBOL</span></span>
<span><span class="va">resU</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tmodUtest.html">tmodUtest</a></span><span class="op">(</span><span class="va">l</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">resU</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                ID                               Title</span></span>
<span><span class="co">## LI.M37.0 LI.M37.0 immune activation - generic cluster</span></span>
<span><span class="co">## DC.M4.2   DC.M4.2                        Inflammation</span></span>
<span><span class="co">## DC.M1.2   DC.M1.2                          Interferon</span></span>
<span><span class="co">## DC.M3.2   DC.M3.2                        Inflammation</span></span>
<span><span class="co">## DC.M5.15 DC.M5.15                         Neutrophils</span></span>
<span><span class="co">## DC.M7.29 DC.M7.29                        Undetermined</span></span>
<span><span class="co">##               U  N1   AUC  P.Value adj.P.Val</span></span>
<span><span class="co">## LI.M37.0 352659 100 0.746 1.60e-17  9.68e-15</span></span>
<span><span class="co">## DC.M4.2   91352  20 0.950 1.67e-12  5.07e-10</span></span>
<span><span class="co">## DC.M1.2   73612  17 0.900 5.70e-09  9.62e-07</span></span>
<span><span class="co">## DC.M3.2   96366  24 0.836 6.35e-09  9.62e-07</span></span>
<span><span class="co">## DC.M5.15  65289  16 0.848 7.24e-07  8.77e-05</span></span>
<span><span class="co">## DC.M7.29  77738  20 0.809 9.08e-07  9.18e-05</span></span></code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">resU</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 39</span></span></code></pre>
<p>This list makes a lot of sense, and also is more stable than the other one: it does not depend on modules that contain just a few genes. Since the statistics is different, the b, B, n, N and E columns in the output have been replaced by the following:</p>
<ul>
<li>
<strong>U</strong> The Mann-Whitney U statistics</li>
<li>
<strong>N1</strong> Number of genes in the module</li>
<li>
<strong>AUC</strong> Area under curve – a measure of the effect size</li>
</ul>
<p>A U-test has been also implemented in limma in the <code><a href="https://rdrr.io/pkg/limma/man/geneSetTest.html" class="external-link">wilcoxGST()</a></code> function.</p>
</div>
<div class="section level4">
<h4 id="cerno-test-tmodcernotest-and-tmodztest">CERNO test (<code>tmodCERNOtest</code> and <code>tmodZtest</code>)<a class="anchor" aria-label="anchor" href="#cerno-test-tmodcernotest-and-tmodztest"></a>
</h4>
<p>There are two tests in tmod which both operate on an ordered list of genes: <code>tmodUtest</code> and <code>tmodCERNOtest</code>. The U test is simple, however has two main issues. Firstly, The CERNO test, described by Yamaguchi et al. <span class="citation">(<a href="#ref-yamaguchi2008ifn" role="doc-biblioref">2008</a>)</span>, is based on Fisher’s method of combining probabilities. In summary, for a given module, the scaled ranks of genes from the module are treated as probabilities. These are then logarithmized, summed and multiplied by -2:</p>
<p><span class="math display">\[f_{CERNO}=-2 \cdot \sum_{i = 1}^{N} \ln{\frac{R_i}{N_{tot}}}\]</span></p>
<p>This statitic has the <span class="math inline">\(\chi^2\)</span> distribution with <span class="math inline">\(2\cdot N\)</span> degrees of freedom, where <span class="math inline">\(N\)</span> is the number of genes in a given module and <span class="math inline">\(N_{tot}\)</span> is the total number of genes <span class="citation">(<a href="#ref-yamaguchi2008ifn" role="doc-biblioref">Yamaguchi et al. 2008</a>)</span>.</p>
<p>The CERNO test is actually much more practical than other tests for most purposes and is the recommended approach. A variant called <code>tmodZtest</code> exists in which the p-values are combined using Stouffer’s method rather than the Fisher’s method.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">l</span>    <span class="op">&lt;-</span> <span class="va">tt</span><span class="op">$</span><span class="va">GENE_SYMBOL</span></span>
<span><span class="va">resCERNO</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tmodUtest.html">tmodCERNOtest</a></span><span class="op">(</span><span class="va">l</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">resCERNO</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                ID                               Title cerno</span></span>
<span><span class="co">## LI.M37.0 LI.M37.0 immune activation - generic cluster   426</span></span>
<span><span class="co">## DC.M4.2   DC.M4.2                        Inflammation   151</span></span>
<span><span class="co">## DC.M3.4   DC.M3.4                          Interferon   129</span></span>
<span><span class="co">## DC.M1.2   DC.M1.2                          Interferon   113</span></span>
<span><span class="co">## DC.M7.29 DC.M7.29                        Undetermined   119</span></span>
<span><span class="co">## LI.M11.0 LI.M11.0          enriched in monocytes (II)   114</span></span>
<span><span class="co">##           N1   AUC  cES  P.Value adj.P.Val</span></span>
<span><span class="co">## LI.M37.0 100 0.746 2.13 1.82e-18  1.11e-15</span></span>
<span><span class="co">## DC.M4.2   20 0.950 3.78 8.04e-15  2.44e-12</span></span>
<span><span class="co">## DC.M3.4   17 0.832 3.81 4.61e-13  9.31e-11</span></span>
<span><span class="co">## DC.M1.2   17 0.900 3.31 2.30e-10  3.48e-08</span></span>
<span><span class="co">## DC.M7.29  20 0.809 2.97 1.00e-09  1.21e-07</span></span>
<span><span class="co">## LI.M11.0  20 0.777 2.85 5.26e-09  5.31e-07</span></span></code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">resCERNO</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 38</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="plage">PLAGE<a class="anchor" aria-label="anchor" href="#plage"></a>
</h4>
<p>PLAGE <span class="citation">(<a href="#ref-tomfohr2005pathway" role="doc-biblioref">Tomfohr, Lu, and Kepler 2005</a>)</span> is a gene set enrichment method based on singular value decomposition (SVD). The idea is that instead of running a statistical test (such as a t-test) on each gene separately, information present in the gene expression of all genes in a gene set is first extracted using SVD, and the resulting vector (one per gene set) is treated as a “pseudo gene” and analysed using the approppriate statistical tool.</p>
<p>In the tmod implementation, for each module a gene expression matrix subset is generated and decomposed using PCA using the <code><a href="../reference/eigengene.html">eigengene()</a></code> function. The first component is returned and a t-test comparing two groups is then performed. This limits the implementation to only two groups, but extending it for more than one group is trivial.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/tmodUtest.html">tmodPLAGEtest</a></span><span class="op">(</span><span class="va">Egambia</span><span class="op">$</span><span class="va">GENE_SYMBOL</span>, <span class="va">Egambia</span><span class="op">[</span>,<span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span><span class="op">]</span>, group<span class="op">=</span><span class="va">group</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Converting group to factor</span></span></code></pre>
<pre><code><span><span class="co">## Calculating eigengenes...</span></span></code></pre>
<pre><code><span><span class="co">##                  ID</span></span>
<span><span class="co">## LI.S4         LI.S4</span></span>
<span><span class="co">## LI.M11.0   LI.M11.0</span></span>
<span><span class="co">## DC.M9.29   DC.M9.29</span></span>
<span><span class="co">## DC.M5.12   DC.M5.12</span></span>
<span><span class="co">## DC.M7.29   DC.M7.29</span></span>
<span><span class="co">## DC.M3.4     DC.M3.4</span></span>
<span><span class="co">## LI.M16       LI.M16</span></span>
<span><span class="co">## DC.M4.2     DC.M4.2</span></span>
<span><span class="co">## DC.M7.16   DC.M7.16</span></span>
<span><span class="co">## LI.M67       LI.M67</span></span>
<span><span class="co">## LI.M37.0   LI.M37.0</span></span>
<span><span class="co">## LI.M4.3     LI.M4.3</span></span>
<span><span class="co">## LI.M118.0 LI.M118.0</span></span>
<span><span class="co">## LI.M37.1   LI.M37.1</span></span>
<span><span class="co">## LI.M112.0 LI.M112.0</span></span>
<span><span class="co">## LI.M105     LI.M105</span></span>
<span><span class="co">## DC.M3.2     DC.M3.2</span></span>
<span><span class="co">## LI.M75       LI.M75</span></span>
<span><span class="co">## LI.M35.0   LI.M35.0</span></span>
<span><span class="co">## LI.M121     LI.M121</span></span>
<span><span class="co">## LI.M165     LI.M165</span></span>
<span><span class="co">## DC.M4.10   DC.M4.10</span></span>
<span><span class="co">## LI.M81       LI.M81</span></span>
<span><span class="co">##                                                      Title</span></span>
<span><span class="co">## LI.S4                           Monocyte surface signature</span></span>
<span><span class="co">## LI.M11.0                        enriched in monocytes (II)</span></span>
<span><span class="co">## DC.M9.29                                      Undetermined</span></span>
<span><span class="co">## DC.M5.12                                        Interferon</span></span>
<span><span class="co">## DC.M7.29                                      Undetermined</span></span>
<span><span class="co">## DC.M3.4                                         Interferon</span></span>
<span><span class="co">## LI.M16                      TLR and inflammatory signaling</span></span>
<span><span class="co">## DC.M4.2                                       Inflammation</span></span>
<span><span class="co">## DC.M7.16                                      Undetermined</span></span>
<span><span class="co">## LI.M67                           activated dendritic cells</span></span>
<span><span class="co">## LI.M37.0               immune activation - generic cluster</span></span>
<span><span class="co">## LI.M4.3   myeloid cell enriched receptors and transporters</span></span>
<span><span class="co">## LI.M118.0                       enriched in monocytes (IV)</span></span>
<span><span class="co">## LI.M37.1                       enriched in neutrophils (I)</span></span>
<span><span class="co">## LI.M112.0                        complement activation (I)</span></span>
<span><span class="co">## LI.M105                                                TBA</span></span>
<span><span class="co">## DC.M3.2                                       Inflammation</span></span>
<span><span class="co">## LI.M75                             antiviral IFN signature</span></span>
<span><span class="co">## LI.M35.0                          signaling in T cells (I)</span></span>
<span><span class="co">## LI.M121                                                TBA</span></span>
<span><span class="co">## LI.M165         enriched in activated dendritic cells (II)</span></span>
<span><span class="co">## DC.M4.10                                            B cell</span></span>
<span><span class="co">## LI.M81             enriched in myeloid cells and monocytes</span></span>
<span><span class="co">##               t     D AbsD  P.Value adj.P.Val</span></span>
<span><span class="co">## LI.S4     -7.17 -2.62 2.62 9.96e-08  5.36e-05</span></span>
<span><span class="co">## LI.M11.0  -6.45 -2.35 2.35 5.51e-07  1.18e-04</span></span>
<span><span class="co">## DC.M9.29  -6.54 -2.39 2.39 6.57e-07  1.18e-04</span></span>
<span><span class="co">## DC.M5.12  -5.97 -2.18 2.18 2.10e-06  2.83e-04</span></span>
<span><span class="co">## DC.M7.29  -5.99 -2.19 2.19 2.92e-06  3.14e-04</span></span>
<span><span class="co">## DC.M3.4   -5.70 -2.08 2.08 4.21e-06  3.78e-04</span></span>
<span><span class="co">## LI.M16    -5.34 -1.95 1.95 1.09e-05  8.41e-04</span></span>
<span><span class="co">## DC.M4.2   -5.22 -1.91 1.91 1.54e-05  1.03e-03</span></span>
<span><span class="co">## DC.M7.16  -5.11 -1.87 1.87 2.18e-05  1.31e-03</span></span>
<span><span class="co">## LI.M67    -4.69 -1.71 1.71 6.59e-05  3.37e-03</span></span>
<span><span class="co">## LI.M37.0  -4.73 -1.73 1.73 6.89e-05  3.37e-03</span></span>
<span><span class="co">## LI.M4.3   -4.63 -1.69 1.69 9.74e-05  3.87e-03</span></span>
<span><span class="co">## LI.M118.0 -4.62 -1.69 1.69 9.75e-05  3.87e-03</span></span>
<span><span class="co">## LI.M37.1  -4.53 -1.66 1.66 1.01e-04  3.87e-03</span></span>
<span><span class="co">## LI.M112.0 -4.36 -1.59 1.59 1.75e-04  6.29e-03</span></span>
<span><span class="co">## LI.M105   -4.10 -1.50 1.50 3.28e-04  1.10e-02</span></span>
<span><span class="co">## DC.M3.2   -3.93 -1.44 1.44 5.15e-04  1.63e-02</span></span>
<span><span class="co">## LI.M75    -3.91 -1.43 1.43 5.63e-04  1.68e-02</span></span>
<span><span class="co">## LI.M35.0  -3.81 -1.39 1.39 7.35e-04  2.00e-02</span></span>
<span><span class="co">## LI.M121   -3.79 -1.39 1.39 7.50e-04  2.00e-02</span></span>
<span><span class="co">## LI.M165   -3.77 -1.38 1.38 7.80e-04  2.00e-02</span></span>
<span><span class="co">## DC.M4.10   3.78  1.38 1.38 8.60e-04  2.10e-02</span></span>
<span><span class="co">## LI.M81    -3.80 -1.39 1.39 9.29e-04  2.17e-02</span></span></code></pre>
</div>
</div>
</div>
<div class="section level2">
<h2 id="visualisation-and-presentation-of-results-in-tmod">Visualisation and presentation of results in tmod<a class="anchor" aria-label="anchor" href="#visualisation-and-presentation-of-results-in-tmod"></a>
</h2>
<div class="section level3">
<h3 id="introduction-3">Introduction<a class="anchor" aria-label="anchor" href="#introduction-3"></a>
</h3>
<p>By default, results produced by tmod are data frames containing one row per tested gene set / module. In certain circumstances, when multiple tests are performed, the returned object is a list in which each element is a results table. In other situations a list can be created manually. In any case, it is often necessary to extract, compare or summarize one or more result tables.</p>
</div>
<div class="section level3">
<h3 id="visualizing-gene-sets-with-eigengene">Visualizing gene sets with eigengene<a class="anchor" aria-label="anchor" href="#visualizing-gene-sets-with-eigengene"></a>
</h3>
<p>One of the most frequent demands is to somehow show differences in transcription in the gene sets between groups. This is usually quite bothersome. One possibility is to use a heatmap, another to show how all genes vary between the groups. Both these options result in messy, frequently biased pictures (for example, if only the “best” genes are selected for a heatmap).</p>
<p>The code below shows how to select genes from a module with the <code><a href="../reference/getModuleMembers.html">getModuleMembers()</a></code> function. For this, we will use the Gambia data set and the LI.M75 interferon module.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m</span> <span class="op">&lt;-</span> <span class="st">"LI.M75"</span></span>
<span><span class="co">## getModuleMembers returns a list – you can choose to </span></span>
<span><span class="co">## select multiple modules</span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getModuleMembers.html">getModuleMembers</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">sel</span> <span class="op">&lt;-</span> <span class="va">Egambia</span><span class="op">$</span><span class="va">GENE_SYMBOL</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">genes</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.matrix.html" class="external-link">data.matrix</a></span><span class="op">(</span><span class="va">Egambia</span><span class="op">)</span><span class="op">[</span><span class="va">sel</span>, <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span><span class="op">]</span> <span class="co"># expression matrix</span></span></code></pre></div>
<p>Matix x contains expression of all genes from the selected module (one row per gene, one column per sample).</p>
<p>Then, heatmaps and line plots can be generated. Note that both of the approaches below are discouraged. Firstly, these representations are chaotic and hard to read. Secondly, it is easy to manipulate such plots in order to make the effects more prominent than they really are. Thirdly, they use up a lot of space and ink to convey very little useful and interpretable information, which is a sin<span class="citation">(<a href="#ref-tufte2014visual" role="doc-biblioref">Tufte and Graves-Morris 2014</a>)</span>.</p>
<p><img src="tmod_files/figure-html/heatmap1-1.png" width="1152"></p>
<p><img src="tmod_files/figure-html/heatmap2-1.png" width="960"></p>
<p>A better (but not perfect) approach is to use <em>eigengenes</em>. An eigengene is a vector of numbers which are thought to represent all genes in a gene set. It is calculated by running a principal component analysis on an expression matrix of the genes of interest. This vector can be thought of as an “average” or “representative” gene of a gene set.</p>
<p>Below, we calculate all eigengenes from modules in tmod and display two of them. The object <code>eig</code> will contain one row per module and one column per sample.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">eig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/eigengene.html">eigengene</a></span><span class="op">(</span><span class="va">Egambia</span><span class="op">[</span>,<span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span><span class="op">]</span>, <span class="va">Egambia</span><span class="op">$</span><span class="va">GENE_SYMBOL</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/showGene.html">showGene</a></span><span class="op">(</span><span class="va">eig</span><span class="op">[</span><span class="st">"LI.M75"</span>, <span class="op">]</span>, <span class="va">group</span>, </span>
<span>  ylab<span class="op">=</span><span class="st">"Eigengene"</span>,</span>
<span>  main<span class="op">=</span><span class="st">"antiviral Interferon signature"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/showGene.html">showGene</a></span><span class="op">(</span><span class="va">eig</span><span class="op">[</span><span class="st">"LI.M16"</span>, <span class="op">]</span>, <span class="va">group</span>, </span>
<span>  ylab<span class="op">=</span><span class="st">"Eigengene"</span>,</span>
<span>  main<span class="op">=</span><span class="st">"TLR and inflammatory signaling"</span><span class="op">)</span></span></code></pre></div>
<p><img src="tmod_files/figure-html/eigengene-1.png" width="768"></p>
<p>In fact, one can compare the eigengenes using a t.test or another statistical procedure – this is the essence of the PLAGE algorithm, described earlier.</p>
</div>
<div class="section level3">
<h3 id="showing-enrichment-with-evidence-plots">Showing enrichment with evidence plots<a class="anchor" aria-label="anchor" href="#showing-enrichment-with-evidence-plots"></a>
</h3>
<p>Let us first investigate in more detail the module LI.M75, the antiviral interferon signature. We can use the <code>ggEvidencePlot</code> function to see how the module is enriched in the list <code>l</code>.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">l</span>    <span class="op">&lt;-</span> <span class="va">tt</span><span class="op">$</span><span class="va">GENE_SYMBOL</span></span>
<span><span class="fu">theme_set</span><span class="op">(</span><span class="fu">theme_minimal</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/ggEvidencePlot.html">ggEvidencePlot</a></span><span class="op">(</span><span class="va">l</span>, <span class="st">"LI.M75"</span><span class="op">)</span> </span></code></pre></div>
<p><img src="tmod_files/figure-html/five-1.png" width="672"></p>
<p>In essence, this is a receiver-operator characteristic (ROC) curve, and the area under the curve (AUC) is related to the U-statistic, from which the P-value in the tmodUtest is calculated, as <span class="math inline">\(\text{AUC}=\frac{U}{n_1\cdot n_2}\)</span>. Both the U statistic and the AUC are reported. Moreover, the AUC can be used to calculate effect size according to the Wendt’s formula<span class="citation">(<a href="#ref-wendt1972dealing" role="doc-biblioref">Wendt 1972</a>)</span> for rank-biserial correlation coefficient:</p>
<p><span class="math display">\[r=1-\frac{2\cdot U}{n_1\cdot n_2} = 1 - 2\cdot\text{AUC}\]</span></p>
<p>In the above diagram, we see that nine out of the 10 genes that belong to the LI.M75 module and which are present in the Egambia data set are ranked among the top 1000 genes (as sorted by p-value).</p>
<p>Evidence plots are an important tool to understand whether the obtained results are not only significant in the statistical sense, but whether they are also of biological interest. As mentioned above, the area under the evidence plot, the AUC (area under curve) is an important measure of the effect size. Effect size should always be taken into consideration when interpreting p-values.</p>
<p>The figure below shows three very different enrichments.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://purrr.tidyverse.org" class="external-link">purrr</a></span><span class="op">)</span></span>
<span><span class="va">sel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"LI.M67"</span>, <span class="st">"LI.M37.0"</span><span class="op">)</span></span>
<span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span><span class="va">sel</span>, <span class="op">~</span> <span class="fu"><a href="../reference/ggEvidencePlot.html">ggEvidencePlot</a></span><span class="op">(</span><span class="va">l</span>, <span class="va">.x</span>, gene.labels<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">plots</span>, labels<span class="op">=</span><span class="va">sel</span><span class="op">)</span></span></code></pre></div>
<p><img src="tmod_files/figure-html/unnamed-chunk-4-1.png" width="1152"></p>
<p>The corresponding results are shown in the following table:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">foo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tmodUtest.html">tmodCERNOtest</a></span><span class="op">(</span><span class="va">l</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">ID</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">sel</span><span class="op">)</span> </span>
<span><span class="va">foo</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="left">ID</th>
<th align="left">Title</th>
<th align="right">cerno</th>
<th align="right">N1</th>
<th align="right">AUC</th>
<th align="right">cES</th>
<th align="right">P.Value</th>
<th align="right">adj.P.Val</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">LI.M37.0</td>
<td align="left">LI.M37.0</td>
<td align="left">immune activation - generic cluster</td>
<td align="right">426.4</td>
<td align="right">100</td>
<td align="right">0.746</td>
<td align="right">2.13</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">LI.M67</td>
<td align="left">LI.M67</td>
<td align="left">activated dendritic cells</td>
<td align="right">49.5</td>
<td align="right">6</td>
<td align="right">0.971</td>
<td align="right">4.13</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
</tbody>
</table>
<p>LI.M67 (activated dendritic cells) contains only 6 genes, but as most of them have really low p-values and are on the top of the p-value sorted list, the resulting effect size is close to 1.</p>
<p>The next panel shows a gene set containing a large number of genes. The effect size for LI.M37.0 is much smaller (0.75), but thanks to the large number of genes the enrichment is significant, with p-value lower than for LI.M67.</p>
<p>Very often, a gene set with a larger number of genes will have a low p-value. However, these gene sets are frequently not very specific, and the corresponding effect sizes are not large. In the above example, the LI.M67 is a much more interesting result: it is more specific and shows a much larger effect, despite having a much higher p-value.</p>
</div>
<div class="section level3">
<h3 id="summary-tables">Summary tables<a class="anchor" aria-label="anchor" href="#summary-tables"></a>
</h3>
<p>We can summarize the output from the previously run tests (<code>tmodUtest</code>, <code>tmodCERNOtest</code> and <code>tmodHGtest</code>) in one table using <code>tmodSummary</code>. For this, we will create a list containing results from all comparisons.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">resAll</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>CERNO<span class="op">=</span><span class="va">resC</span>, U<span class="op">=</span><span class="va">resU</span>, HG<span class="op">=</span><span class="va">resHG</span><span class="op">)</span></span>
<span><span class="co">#head(tmodSummary(resAll))</span></span></code></pre></div>
<p>The table below shows the results.</p>
<table class="table">
<colgroup>
<col width="11%">
<col width="20%">
<col width="12%">
<col width="12%">
<col width="9%">
<col width="12%">
<col width="8%">
<col width="12%">
</colgroup>
<thead><tr class="header">
<th align="left">ID</th>
<th align="left">Title</th>
<th align="right">AUC.CERNO</th>
<th align="right">q.CERNO</th>
<th align="right">AUC.U</th>
<th align="right">q.U</th>
<th align="right">E.HG</th>
<th align="right">q.HG</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">DC.M1.2</td>
<td align="left">Interferon</td>
<td align="right">0.9004</td>
<td align="right">3.482e-08</td>
<td align="right">0.9004</td>
<td align="right">9.624e-07</td>
<td align="right">18.12</td>
<td align="right">0.02948</td>
</tr>
<tr class="even">
<td align="left">DC.M3.2</td>
<td align="left">Inflammation</td>
<td align="right">0.8362</td>
<td align="right">1.05e-06</td>
<td align="right">0.8362</td>
<td align="right">9.624e-07</td>
<td align="right">NA</td>
<td align="right">NA</td>
</tr>
<tr class="odd">
<td align="left">DC.M3.4</td>
<td align="left">Interferon</td>
<td align="right">0.8316</td>
<td align="right">9.311e-11</td>
<td align="right">0.8316</td>
<td align="right">9.865e-05</td>
<td align="right">42.28</td>
<td align="right">5.721e-08</td>
</tr>
<tr class="even">
<td align="left">DC.M3.6</td>
<td align="left">Cytotoxic/NK Cell</td>
<td align="right">0.7933</td>
<td align="right">0.03038</td>
<td align="right">0.7933</td>
<td align="right">0.01008</td>
<td align="right">NA</td>
<td align="right">NA</td>
</tr>
<tr class="odd">
<td align="left">DC.M4.10</td>
<td align="left">B cell</td>
<td align="right">0.9076</td>
<td align="right">0.004302</td>
<td align="right">0.9076</td>
<td align="right">0.003376</td>
<td align="right">NA</td>
<td align="right">NA</td>
</tr>
<tr class="even">
<td align="left">DC.M4.2</td>
<td align="left">Inflammation</td>
<td align="right">0.9504</td>
<td align="right">2.436e-12</td>
<td align="right">0.9504</td>
<td align="right">5.075e-10</td>
<td align="right">30.8</td>
<td align="right">6.507e-06</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="panel-plots">Panel plots<a class="anchor" aria-label="anchor" href="#panel-plots"></a>
</h3>
<p>A list of result tables (or even of a single result table) can be visualized using a heatmap-like plot called here “panel plot.” The idea is to show both, effect sizes and p-values, and, optionally, also the direction of gene regulation.</p>
<p>In the example below, we will use the <code>resAll</code> object created above, containing the results from three different tests for enrichment, to compare the results of the individual tests. However, since the <code>E</code> column of HG test is not easily comparable to the AUC values (which are between 0 and 1), we need to scale it down. Also, we need to call it “AUC,” because otherwise we can’t show the values on the same plot.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">resAll</span><span class="op">$</span><span class="va">HG</span><span class="op">$</span><span class="va">AUC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">resAll</span><span class="op">$</span><span class="va">HG</span><span class="op">$</span><span class="va">E</span><span class="op">)</span> <span class="op">-</span> <span class="fl">0.5</span></span>
<span><span class="fu"><a href="../reference/ggPanelplot.html">ggPanelplot</a></span><span class="op">(</span><span class="va">resAll</span><span class="op">)</span></span></code></pre></div>
<p><img src="tmod_files/figure-html/panelplots-1.png" width="768"></p>
<p>Each enrichment result corresponds to a reddish bar. The size of the bar corresponds to the effect size (AUC or log10(Enrichment) - 0.5, as it may be), and color intensity corresponds to the p-value – pale colors show p-values closer to 0.01. It is easily seen how <code>tmodCERNOtest</code> is the more sensitive option.</p>
<p>We can see that also the intercept term is enriched for genes found in monocytes and neutrophils. Note that by default, ggPanelplot only shows enrichments with p &lt; 0.01, hence a slight difference from the tmodSummary output. This behavior can be modified by the <code>q_thr</code> option:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ggPanelplot.html">ggPanelplot</a></span><span class="op">(</span><span class="va">resAll</span>, q_thr<span class="op">=</span><span class="fl">1e-3</span><span class="op">)</span></span></code></pre></div>
<p><img src="tmod_files/figure-html/panelplots2-1.png" width="768"></p>
<p>However, one is usually interested in the direction of regulation. If a gene list is sorted by p-value, the enriched modules may contain both up- or down-regulated genes<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;Searching for enrichment only in up- or only in down-regulated genes depends on how the gene list is sorted; this is described in Section “&lt;a href="#updown"&gt;Testing for up- or down-regulated genes separately&lt;/a&gt;.”&lt;/p&gt;'><sup>4</sup></a>. It is often desirable to visualize whether genes in a module go up, or go down between experimental conditions. For this, the function <code>tmodDecideTests</code> is used to obtain the number of significantly up- or down-regulated genes in a module.</p>
<p>This information must be obtained separately from the differential gene expression analysis and provided as a list to <code>ggPanelplot</code>. The names of this list must be identical to the names in the results list. Below, we reuse the same object for all three tests, since the DEGs (differentially expressed genes) were the same in all three comparisons.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">degs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tmodDecideTests.html">tmodDecideTests</a></span><span class="op">(</span>g<span class="op">=</span><span class="va">tt</span><span class="op">$</span><span class="va">GENE_SYMBOL</span>, lfc<span class="op">=</span><span class="va">tt</span><span class="op">$</span><span class="va">logFC</span>,</span>
<span>                       pval<span class="op">=</span><span class="va">tt</span><span class="op">$</span><span class="va">adj.P.Val</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">degs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>CERNO<span class="op">=</span><span class="va">degs</span>, HG<span class="op">=</span><span class="va">degs</span>, U<span class="op">=</span><span class="va">degs</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/ggPanelplot.html">ggPanelplot</a></span><span class="op">(</span><span class="va">resAll</span>, sgenes <span class="op">=</span> <span class="va">degs</span><span class="op">)</span></span></code></pre></div>
<p><img src="tmod_files/figure-html/pie-1.png" width="960"></p>
<p>Each mini-plot shows the effect size of the enrichment and the corresponding p-value, as before. Additionally, the fraction of up-regulated and down-regulated genes is visualized by coloring a fraction of the area of the mini-plot red or blue, respectively<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;The colors can be modified by the parameter &lt;code&gt;colors&lt;/code&gt;&lt;/p&gt;"><sup>5</sup></a>.</p>
<p>The <code>ggPanelplot</code> function has several parameters, notably for filtering and labelling:</p>
<ul>
<li>Filtering:
<ul>
<li>
<code>filter_row_q</code> and <code>filter_row_auc</code> remove gene sets for which, respectively, the FDR p-value or the effect size are not achieved in any of the comparisons;</li>
<li>
<code>q_cutoff</code>: enrichments with q-values above this cutoff are considered to be absent</li>
</ul>
</li>
</ul>
<p>The <code>ggPanelplot</code> function returns a ggplot2 graph, and therefore allows much customization.</p>
</div>
</div>
<div class="section level2">
<h2 id="using-tmod-for-other-types-of-gse-analyses">Using tmod for other types of GSE analyses<a class="anchor" aria-label="anchor" href="#using-tmod-for-other-types-of-gse-analyses"></a>
</h2>
<p>The fact that tmod relies on a single ordered list of genes makes it useful in many other situations in which such a list presents itself.</p>
<div class="section level3">
<h3 id="correlation-analysis">Correlation analysis<a class="anchor" aria-label="anchor" href="#correlation-analysis"></a>
</h3>
<p>Genes can be ordered by their absolute correlation with a variable or even a data set or a module. For example, one can ask the question about a function of a particular unknown gene – such as ANKRD22, annotated as “ankyrin repeat domain 22.”</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="va">E</span><span class="op">[</span> <span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="st">"ANKRD22"</span>, <span class="va">Egambia</span><span class="op">$</span><span class="va">GENE_SYMBOL</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">cors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">x</span>, <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">E</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">ord</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">cors</span><span class="op">)</span>, decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/tmodUtest.html">tmodCERNOtest</a></span><span class="op">(</span><span class="va">Egambia</span><span class="op">$</span><span class="va">GENE_SYMBOL</span><span class="op">[</span> <span class="va">ord</span> <span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                ID</span></span>
<span><span class="co">## LI.M37.0 LI.M37.0</span></span>
<span><span class="co">## DC.M3.4   DC.M3.4</span></span>
<span><span class="co">## DC.M4.2   DC.M4.2</span></span>
<span><span class="co">## DC.M1.2   DC.M1.2</span></span>
<span><span class="co">## DC.M7.29 DC.M7.29</span></span>
<span><span class="co">## LI.M165   LI.M165</span></span>
<span><span class="co">##                                               Title cerno</span></span>
<span><span class="co">## LI.M37.0        immune activation - generic cluster   431</span></span>
<span><span class="co">## DC.M3.4                                  Interferon   142</span></span>
<span><span class="co">## DC.M4.2                                Inflammation   151</span></span>
<span><span class="co">## DC.M1.2                                  Interferon   132</span></span>
<span><span class="co">## DC.M7.29                               Undetermined   117</span></span>
<span><span class="co">## LI.M165  enriched in activated dendritic cells (II)   113</span></span>
<span><span class="co">##           N1   AUC  cES  P.Value adj.P.Val</span></span>
<span><span class="co">## LI.M37.0 100 0.719 2.16 4.71e-19  2.85e-16</span></span>
<span><span class="co">## DC.M3.4   17 0.866 4.17 4.25e-15  1.29e-12</span></span>
<span><span class="co">## DC.M4.2   20 0.913 3.79 7.25e-15  1.46e-12</span></span>
<span><span class="co">## DC.M1.2   17 0.926 3.89 1.68e-13  2.54e-11</span></span>
<span><span class="co">## DC.M7.29  20 0.808 2.92 1.83e-09  2.20e-07</span></span>
<span><span class="co">## LI.M165   19 0.781 2.98 2.18e-09  2.20e-07</span></span></code></pre>
<p>Clearly, ANKRD22 correlates to other immune related genes, most of all these which are interferon inducible.</p>
<p>In another example, consider correlation between genes and the first principal component (“eigengene”) of a group of genes of unknown function<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;More on eigengenes in the Chapter on modules&lt;/p&gt;"><sup>6</sup></a>. To demonstrate the method, we will select the genes from the module “LI.M75.” For this, we use the function <code>getGenes</code> with the optional argument <code>genes</code> used to filter the genes in the module by the genes present in the data set.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getGenes.html">getGenes</a></span><span class="op">(</span><span class="st">"LI.M75"</span>, genes<span class="op">=</span><span class="va">Egambia</span><span class="op">$</span><span class="va">GENE_SYMBOL</span>, </span>
<span>              as.list<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="va">E</span><span class="op">[</span> <span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">g</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="va">Egambia</span><span class="op">$</span><span class="va">GENE_SYMBOL</span><span class="op">)</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co">## calculating the "eigengene" (PC1)</span></span>
<span><span class="va">pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html" class="external-link">prcomp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>, scale.<span class="op">=</span><span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">eigen</span> <span class="op">&lt;-</span> <span class="va">pca</span><span class="op">$</span><span class="va">x</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">cors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">eigen</span>, <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">E</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## order all genes by the correlation between the gene and the PC1</span></span>
<span><span class="va">ord</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">cors</span><span class="op">)</span>, decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/tmodUtest.html">tmodCERNOtest</a></span><span class="op">(</span><span class="va">Egambia</span><span class="op">$</span><span class="va">GENE_SYMBOL</span><span class="op">[</span> <span class="va">ord</span> <span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                ID</span></span>
<span><span class="co">## DC.M1.2   DC.M1.2</span></span>
<span><span class="co">## DC.M3.4   DC.M3.4</span></span>
<span><span class="co">## LI.M165   LI.M165</span></span>
<span><span class="co">## LI.M75     LI.M75</span></span>
<span><span class="co">## LI.M37.0 LI.M37.0</span></span>
<span><span class="co">## DC.M4.2   DC.M4.2</span></span>
<span><span class="co">##                                               Title cerno</span></span>
<span><span class="co">## DC.M1.2                                  Interferon   197</span></span>
<span><span class="co">## DC.M3.4                                  Interferon   154</span></span>
<span><span class="co">## LI.M165  enriched in activated dendritic cells (II)   155</span></span>
<span><span class="co">## LI.M75                      antiviral IFN signature   104</span></span>
<span><span class="co">## LI.M37.0        immune activation - generic cluster   364</span></span>
<span><span class="co">## DC.M4.2                                Inflammation   128</span></span>
<span><span class="co">##           N1   AUC  cES  P.Value adj.P.Val</span></span>
<span><span class="co">## DC.M1.2   17 0.964 5.80 6.77e-25  4.10e-22</span></span>
<span><span class="co">## DC.M3.4   17 0.926 4.53 3.10e-17  9.40e-15</span></span>
<span><span class="co">## LI.M165   19 0.825 4.07 5.46e-16  1.10e-13</span></span>
<span><span class="co">## LI.M75    10 0.943 5.21 2.25e-13  3.42e-11</span></span>
<span><span class="co">## LI.M37.0 100 0.668 1.82 1.12e-11  1.36e-09</span></span>
<span><span class="co">## DC.M4.2   20 0.888 3.19 4.44e-11  4.49e-09</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="functional-multivariate-analysis">Functional multivariate analysis<a class="anchor" aria-label="anchor" href="#functional-multivariate-analysis"></a>
</h3>
<p>Transcriptional modules can help to understand the biological meaning of the calculated multivariate transformations. For example, consider a principal component analysis (PCA), visualised using the pca3d package <span class="citation">(<a href="#ref-pca3d" role="doc-biblioref">Weiner 2013</a>)</span>:</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mypal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#E69F00"</span>, <span class="st">"#56B4E9"</span><span class="op">)</span></span>
<span><span class="va">pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html" class="external-link">prcomp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">E</span><span class="op">)</span>, scale.<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">col</span> <span class="op">&lt;-</span> <span class="va">mypal</span><span class="op">[</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">group</span><span class="op">)</span> <span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">l</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/pcaplot.html">pcaplot</a></span><span class="op">(</span><span class="va">pca</span>, group<span class="op">=</span><span class="va">group</span>, col<span class="op">=</span><span class="va">col</span><span class="op">)</span></span>
<span> </span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="st">"topleft"</span>, <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">l</span><span class="op">$</span><span class="va">groups</span><span class="op">)</span>,</span>
<span>       pch<span class="op">=</span><span class="va">l</span><span class="op">$</span><span class="va">pch</span>,</span>
<span>       col<span class="op">=</span><span class="va">l</span><span class="op">$</span><span class="va">colors</span>, bty<span class="op">=</span><span class="st">"n"</span><span class="op">)</span></span>
<span><span class="va">l</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/pcaplot.html">pcaplot</a></span><span class="op">(</span><span class="va">pca</span>, group<span class="op">=</span><span class="va">group</span>, col<span class="op">=</span><span class="va">col</span>, components<span class="op">=</span><span class="fl">3</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="st">"topleft"</span>, <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">l</span><span class="op">$</span><span class="va">groups</span><span class="op">)</span>,</span>
<span>       pch<span class="op">=</span><span class="va">l</span><span class="op">$</span><span class="va">pch</span>,</span>
<span>       col<span class="op">=</span><span class="va">l</span><span class="op">$</span><span class="va">colors</span>, bty<span class="op">=</span><span class="st">"n"</span><span class="op">)</span></span></code></pre></div>
<p><img src="tmod_files/figure-html/six-1.png" width="768"></p>
<p>The fourth component looks really interesting. Does it correspond to the modules which we have found before? Each principal component is, after all, a linear combination of gene expression values multiplied by weights (or scores) which are constant for a given component. The <em>i</em>-th principal component for sample <em>j</em> is given by</p>
<p><span class="math display">\[PC_{i,j} = \sum_{k} w_{i,k} \cdot x_{k,j}\]</span></p>
<p>where <span class="math inline">\(k\)</span> is the index of the variables (genes in our case), <span class="math inline">\(w_{i,k}\)</span> is the weight associated with the <span class="math inline">\(i\)</span>-th component and the <span class="math inline">\(k\)</span>-th variable (gene), and <span class="math inline">\(x_{k,j}\)</span> is the value of the variable <span class="math inline">\(k\)</span> for the sample <span class="math inline">\(j\)</span>; that is, the gene expression of gene <span class="math inline">\(k\)</span> in the sample <span class="math inline">\(j\)</span>. Genes influence the position of a sample along a given component the more the larger their absolute weight for that component.</p>
<p>For example, on the right-hand figure above, we see that samples which were taken from TB patients have a high value of the principal component 4; the opposite is true for the healthy controls. The genes that allow us to differentiate between these two groups will have very large, positive weights for genes highly expressed in TB patients, and very large, negative weights for genes which are highly expressed in NID, but not TB.</p>
<p>We can sort the genes by their weight in the given component, since the weights are stored in the pca object in the “rotation” slot, and use the tmodUtest function to test for enrichment of the modules.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">o</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">pca</span><span class="op">$</span><span class="va">rotation</span><span class="op">[</span>,<span class="fl">4</span><span class="op">]</span><span class="op">)</span>, decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">l</span> <span class="op">&lt;-</span> <span class="va">Egambia</span><span class="op">$</span><span class="va">GENE_SYMBOL</span><span class="op">[</span><span class="va">o</span><span class="op">]</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tmodUtest.html">tmodUtest</a></span><span class="op">(</span><span class="va">l</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                ID                               Title</span></span>
<span><span class="co">## LI.M37.0 LI.M37.0 immune activation - generic cluster</span></span>
<span><span class="co">## DC.M4.2   DC.M4.2                        Inflammation</span></span>
<span><span class="co">## DC.M1.2   DC.M1.2                          Interferon</span></span>
<span><span class="co">## DC.M3.2   DC.M3.2                        Inflammation</span></span>
<span><span class="co">## DC.M7.29 DC.M7.29                        Undetermined</span></span>
<span><span class="co">## DC.M3.4   DC.M3.4                          Interferon</span></span>
<span><span class="co">##               U  N1   AUC  P.Value adj.P.Val</span></span>
<span><span class="co">## LI.M37.0 339742 100 0.719 3.13e-14  1.90e-11</span></span>
<span><span class="co">## DC.M4.2   89378  20 0.930 1.52e-11  4.60e-09</span></span>
<span><span class="co">## DC.M1.2   74828  17 0.915 1.61e-09  3.24e-07</span></span>
<span><span class="co">## DC.M3.2   95685  24 0.830 1.14e-08  1.72e-06</span></span>
<span><span class="co">## DC.M7.29  78752  20 0.819 3.99e-07  4.84e-05</span></span>
<span><span class="co">## DC.M3.4   68058  17 0.832 1.07e-06  1.08e-04</span></span></code></pre>
<p>Perfect, this is what we expected: we see that the neutrophil / interferon signature which is the hallmark of the TB biosignature. What about other components? We can run the enrichment for each component and visualise the results using tmod’s functions tmodSummary and ggPanelplot. Below, we use the filter.empty option to omit the principal components which show no enrichment at all.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Calculate enrichment for each component</span></span>
<span><span class="va">gs</span>   <span class="op">&lt;-</span> <span class="va">Egambia</span><span class="op">$</span><span class="va">GENE_SYMBOL</span></span>
<span><span class="co"># function calculating the enrichment of a PC</span></span>
<span><span class="va">gn.f</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">r</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="../reference/tmodUtest.html">tmodCERNOtest</a></span><span class="op">(</span><span class="va">gs</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">r</span><span class="op">)</span>, decreasing<span class="op">=</span><span class="cn">T</span><span class="op">)</span><span class="op">]</span>,</span>
<span>                qval<span class="op">=</span><span class="fl">0.01</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">pca</span><span class="op">$</span><span class="va">rotation</span>, <span class="fl">2</span>, <span class="va">gn.f</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/tmodSummary.html">tmodSummary</a></span><span class="op">(</span><span class="va">x</span>, filter.empty<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##              ID        Title AUC.PC3    q.PC3 AUC.PC4</span></span>
<span><span class="co">## DC.M1.1 DC.M1.1    Platelets      NA       NA      NA</span></span>
<span><span class="co">## DC.M1.2 DC.M1.2   Interferon      NA       NA   0.915</span></span>
<span><span class="co">## DC.M2.3 DC.M2.3 Erythrocytes   0.897 6.01e-12      NA</span></span>
<span><span class="co">## DC.M3.1 DC.M3.1 Erythrocytes   0.729 1.92e-03      NA</span></span>
<span><span class="co">## DC.M3.2 DC.M3.2 Inflammation   0.715 8.88e-03   0.830</span></span>
<span><span class="co">##            q.PC4 AUC.PC9 q.PC9 AUC.PC14   q.PC14 AUC.PC30</span></span>
<span><span class="co">## DC.M1.1       NA      NA    NA    0.746 7.94e-05       NA</span></span>
<span><span class="co">## DC.M1.2 1.75e-08      NA    NA       NA       NA    0.828</span></span>
<span><span class="co">## DC.M2.3       NA      NA    NA       NA       NA    0.843</span></span>
<span><span class="co">## DC.M3.1       NA      NA    NA       NA       NA       NA</span></span>
<span><span class="co">## DC.M3.2 7.16e-10      NA    NA       NA       NA    0.758</span></span>
<span><span class="co">##           q.PC30</span></span>
<span><span class="co">## DC.M1.1       NA</span></span>
<span><span class="co">## DC.M1.2 6.79e-03</span></span>
<span><span class="co">## DC.M2.3 3.08e-07</span></span>
<span><span class="co">## DC.M3.1       NA</span></span>
<span><span class="co">## DC.M3.2 1.17e-05</span></span></code></pre>
<p>The following plot shows the same information in a visual form. The size of the blobs corresponds to the effect size (AUC value), and their color – to the q-value.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ggPanelplot.html">ggPanelplot</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span></code></pre></div>
<p><img src="tmod_files/figure-html/pcsum2-1.png" width="768"></p>
<p>However, we might want to ask, for each module, how many of the genes in that module have a negative, and how many have a positive weight? We can use the function tmodDecideTests for that. For each principal component shown, we want to know how many genes have very large (in absolute terms) weights – we can use the “lfc” parameter of tmodDecideTests for that. We define here “large” as being in the top 25% of all weights in the given component. For this, we need first to calculate the 3rd quartile (top 25% threshold). We will show only 10 components:</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qfnc</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">r</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">r</span>, <span class="fl">0.75</span><span class="op">)</span></span>
<span><span class="va">qqs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">pca</span><span class="op">$</span><span class="va">rotation</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span>, <span class="fl">2</span>, <span class="va">qfnc</span><span class="op">)</span></span>
<span><span class="va">gloadings</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tmodDecideTests.html">tmodDecideTests</a></span><span class="op">(</span><span class="va">gs</span>, lfc<span class="op">=</span><span class="va">pca</span><span class="op">$</span><span class="va">rotation</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span>, lfc.thr<span class="op">=</span><span class="va">qqs</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/ggPanelplot.html">ggPanelplot</a></span><span class="op">(</span><span class="va">x</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span>, sgenes <span class="op">=</span> <span class="va">gloadings</span><span class="op">)</span> </span></code></pre></div>
<p><img src="tmod_files/figure-html/pcsum3-1.png" width="768"></p>
</div>
</div>
<div class="section level2">
<h2 id="using-and-creating-modules-and-gene-sets">Using and creating modules and gene sets<a class="anchor" aria-label="anchor" href="#using-and-creating-modules-and-gene-sets"></a>
</h2>
<p>Tmod was created with transcriptional modules in mind. This is why the word “module” is used throughout tmod. However, any gene or variable set – depending on application – is a “module” in tmod. These data sets can be used with most of tmod functions (including the gene set enrichment test functions) by specifying it with the option <code>mset=</code>, for example <code>tmodCERNOtest(..., mset=mytmodobject)</code>.</p>
<div class="section level3">
<h3 id="using-built-in-gene-sets-transcriptional-modules">Using built-in gene sets (transcriptional modules)<a class="anchor" aria-label="anchor" href="#using-built-in-gene-sets-transcriptional-modules"></a>
</h3>
<p>By default, tmod uses both the modules published by Li et al. <span class="citation">(<a href="#ref-li2014molecular" role="doc-biblioref">Li et al. 2014</a>)</span> (LI) and second set of modules published by Chaussabel et al. <span class="citation">(<a href="#ref-chaussabel2008modular" role="doc-biblioref">Chaussabel et al. 2008</a>)</span> (DC). The module definitions for the DC set were described by Banchereau et al. <span class="citation">(<a href="#ref-banchereau2012host" role="doc-biblioref">Banchereau et al. 2012</a>)</span>.</p>
<p>Depending on the <code>mset</code> parameter to the test functions, either the LI or DC sets are used, or both, if the <code>mset=all</code> has been specified.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">l</span>    <span class="op">&lt;-</span> <span class="va">tt</span><span class="op">$</span><span class="va">GENE_SYMBOL</span></span>
<span><span class="va">res2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tmodUtest.html">tmodUtest</a></span><span class="op">(</span><span class="va">l</span>, mset<span class="op">=</span><span class="st">"LI"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span> <span class="va">res2</span> <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                ID                               Title</span></span>
<span><span class="co">## LI.M37.0 LI.M37.0 immune activation - generic cluster</span></span>
<span><span class="co">## LI.M37.1 LI.M37.1         enriched in neutrophils (I)</span></span>
<span><span class="co">## LI.S4       LI.S4          Monocyte surface signature</span></span>
<span><span class="co">## LI.M75     LI.M75             antiviral IFN signature</span></span>
<span><span class="co">## LI.M11.0 LI.M11.0          enriched in monocytes (II)</span></span>
<span><span class="co">## LI.M67     LI.M67           activated dendritic cells</span></span>
<span><span class="co">##               U  N1   AUC  P.Value adj.P.Val</span></span>
<span><span class="co">## LI.M37.0 352659 100 0.746 1.60e-17  5.53e-15</span></span>
<span><span class="co">## LI.M37.1  50280  12 0.870 4.53e-06  6.57e-04</span></span>
<span><span class="co">## LI.S4     43220  10 0.897 6.85e-06  6.57e-04</span></span>
<span><span class="co">## LI.M75    42996  10 0.893 8.63e-06  6.57e-04</span></span>
<span><span class="co">## LI.M11.0  74652  20 0.777 9.49e-06  6.57e-04</span></span>
<span><span class="co">## LI.M67    28095   6 0.971 3.20e-05  1.81e-03</span></span></code></pre>
<p>As you can see, the information contained in both module sets is partially redundant.</p>
</div>
<div class="section level3">
<h3 id="accessing-the-tmod-gene-set-data">Accessing the tmod gene set data<a class="anchor" aria-label="anchor" href="#accessing-the-tmod-gene-set-data"></a>
</h3>
<p>The <code>tmod</code> package stores its data in two data frames and two lists. This object is contained in a list called <code>tmod</code>, which is loaded with <code>data("tmod")</code>. The names mimick the various environments from Annotation.dbi packages, but currently the objects are just two lists and two data frames.</p>
<ul>
<li>
<strong><code>tmod$gs</code></strong> is a data frame which contains general module information as defined in the supplementary materials for Li et al. <span class="citation">(<a href="#ref-li2014molecular" role="doc-biblioref">Li et al. 2014</a>)</span> and Chaussabel et al. <span class="citation">(<a href="#ref-chaussabel2008modular" role="doc-biblioref">Chaussabel et al. 2008</a>)</span>
</li>
<li>
<strong><code>tmod$gv</code></strong> is vector with gene identifiers;</li>
<li>
<strong><code>tmod$gs2gv</code></strong> contains the mapping between gene sets and genes.</li>
</ul>
<div class="section level4">
<h4 id="module-operations">Module operations<a class="anchor" aria-label="anchor" href="#module-operations"></a>
</h4>
<p>The gene sets used by tmod are objects of class <code>tmod</code>. The default object used in the gene set enrichment tests in the tmod package can be loaded into the environment with the command <code>data(tmod)</code>:</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">tmod</span><span class="op">)</span></span>
<span><span class="va">tmod</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class "tmodGS"</span></span>
<span><span class="co">##  606 gene sets, 12712 genes</span></span></code></pre>
<p>Objects of the class tmod can be easily generated from a number of data sources (see below). Several functions can be used on the objects:</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">tmod</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 606</span></span></code></pre>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"Interferon"</span>, <span class="va">tmod</span><span class="op">$</span><span class="va">gs</span><span class="op">$</span><span class="va">Title</span>, ignore.case<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">ifn</span> <span class="op">&lt;-</span> <span class="va">tmod</span><span class="op">[</span><span class="va">sel</span><span class="op">]</span></span>
<span><span class="va">ifn</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class "tmodGS"</span></span>
<span><span class="co">##  6 gene sets, 161 genes</span></span></code></pre>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">ifn</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 6</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="custom-module-definitions">Custom module definitions<a class="anchor" aria-label="anchor" href="#custom-module-definitions"></a>
</h4>
<p>It is possible to use any kind of arbitrary or custom gene set definitions. These custom definition of gene sets takes form of a list which is then provided as the <code>mset</code> parameter to the test functions. The list in question must have the following members:</p>
<ul>
<li>
<strong>gs</strong> (gene sets) A data frame which contains at least the columns “ID” and “Title.”</li>
<li>
<strong>gs2gv</strong> A list. Mapping between gene sets and the gene vector. Each element is an <em>integer</em> vector which contains the positions of the given gene in the <em>gv</em> vector. The names of the list correspond to the <strong>gs$ID</strong> vector.</li>
<li>
<strong>gv</strong> Gene vector. Character vector of genes.</li>
</ul>
<p>The tests in the tmod package will accept a simple list that contains the above fields. However, the function <code>makeTmodGS</code> can be used conveniently to create a tmod object. This function requires only two parameters: <code>gs</code> – a data frame, as described above – and a mapping between gene sets <em>and gene identifiers</em>, in the parameter <code>gs2gene</code>.</p>
<p>Here is a minimal definition of such a set:</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mymset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tmodGS-class.html">makeTmodGS</a></span><span class="op">(</span></span>
<span>  gs<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>ID<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span>,</span>
<span>             Title<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A title"</span>, </span>
<span>                      <span class="st">"B title"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  gs2gene<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    A<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"G1"</span>, <span class="st">"G2"</span><span class="op">)</span>,</span>
<span>    B<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"G3"</span>, <span class="st">"G4"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">mymset</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class "tmodGS"</span></span>
<span><span class="co">##  2 gene sets, 4 genes</span></span></code></pre>
<p>Whether the gene IDs are Entrez, or something else entirely does not matter, as long as they matched the provided input to the test functions.</p>
</div>
</div>
<div class="section level3">
<h3 id="obtaining-other-gene-sets">Obtaining other gene sets<a class="anchor" aria-label="anchor" href="#obtaining-other-gene-sets"></a>
</h3>
<p>The tests in the tmod package can take any arbitrary module definitions. While tmod – for many reasons – cannot distribute all module sets, it can easily import gene sets from many sources. A few of these will be discussed below.</p>
<div class="section level4">
<h4 id="msigdb">MSigDB<a class="anchor" aria-label="anchor" href="#msigdb"></a>
</h4>
<p>The MSigDB database from the Broad institute is an interesting collection of gene sets (actually, multiple collections), including Reactome pathways, gene ontologies (GO) and many other data sets. Moreover, it is the basis for the GSEA program.</p>
<p>The whole MSigDB is provided by the <a href="https://CRAN.R-project.org/package=msigdbr" class="external-link">msigdbr</a> package from BioConductor. We can then use tmod function <code>makeTmodFromDataFrame</code> to convert the msigdbr data frame into one large tmod object:</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://igordot.github.io/msigdbr/" class="external-link">msigdbr</a></span><span class="op">)</span></span>
<span><span class="va">msig</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://igordot.github.io/msigdbr/reference/msigdbr.html" class="external-link">msigdbr</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">msig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeTmodFromDataFrame.html">makeTmodFromDataFrame</a></span><span class="op">(</span>df<span class="op">=</span><span class="va">msig</span>, </span>
<span>  feature_col<span class="op">=</span><span class="st">"gene_symbol"</span>,</span>
<span>  module_col<span class="op">=</span><span class="st">"gs_id"</span>, title_col<span class="op">=</span><span class="st">"gs_name"</span>, </span>
<span>  extra_module_cols<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gs_cat"</span>, <span class="st">"gs_subcat"</span>, <span class="st">"gs_url"</span>, </span>
<span>                      <span class="st">"gs_exact_source"</span>, <span class="st">"gs_description"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Alternatively, you can the MSigDB in XML format<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;Note that even if you register with MSig, it is not possible to download the database directly from R in the XML format.&lt;/p&gt;"><sup>7</sup></a>. This file can be accessed at the from the pages of MSigDB <a href="http://www.gsea-msigdb.org/gsea/msigdb/download_file.jsp?filePath=/msigdb/release/2022.1.Hs/msigdb_v2022.1.Hs_chip_files_to_download_locally.zip" class="external-link">at the Broad Institute</a> <a href="http://software.broadinstitute.org/gsea/msigdb/download_file.jsp?filePath=/resources/msigdb/6.1/msigdb_v6.1.xml" class="external-link uri">http://software.broadinstitute.org/gsea/msigdb/download_file.jsp?filePath=/resources/msigdb/6.1/msigdb_v6.1.xml</a> – follow the link, register and log in, and save the zip archive on your disk (roughly 113 MB). The ZIP file contains an XML file (called ‘msigdb_v2022.1.Hs.xml’ at the time of writing) which you can import into tmod.</p>
<p>Importing MSigDB from XML is easy – tmod has a function specifically for that purpose. Once you have downloaded the MSigDB file, you can create the tmod-compatible R object with one command<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;MSigDB gene sets can be also downloaded as “GMT” files. This format contains less information and is therefore less usable.&lt;/p&gt;"><sup>8</sup></a>. However, the tmod function tmodImportMsigDB() can also use this format, look up the manual page:</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">msig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tmodImportMSigDB.html">tmodImportMSigDB</a></span><span class="op">(</span><span class="st">"msigdb_v2022.1.Hs.xml"</span><span class="op">)</span></span></code></pre></div>
<p>That’s it – now you can use the full MSigDB for enrichment tests:</p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tmodUtest.html">tmodCERNOtest</a></span><span class="op">(</span><span class="va">tt</span><span class="op">$</span><span class="va">GENE_SYMBOL</span>, mset<span class="op">=</span><span class="va">msig</span><span class="op">)</span></span></code></pre></div>
<p>The results are quite typical for MSigDB, which is quite abundant with similar or overlapping gene sets. As the first results, we see, again, interferon response, as well as sets of genes which are significantly upregulated after yellow fever vaccination – and which are also interferon related. We might want to limit our analysis only to the 50 “hallmark” module categories:</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sel</span> <span class="op">&lt;-</span> <span class="va">msig</span><span class="op">$</span><span class="va">gs</span><span class="op">$</span><span class="va">gs_cat</span> <span class="op">==</span> <span class="st">"H"</span></span>
<span><span class="fu"><a href="../reference/tmodUtest.html">tmodCERNOtest</a></span><span class="op">(</span><span class="va">tt</span><span class="op">$</span><span class="va">GENE_SYMBOL</span>, mset<span class="op">=</span><span class="va">msig</span><span class="op">[</span><span class="va">sel</span><span class="op">]</span> <span class="op">)</span></span></code></pre></div>
<p>Other particularly interesting subsets of MSigDB are shown in the table below. “Category” and “Subcategory” are columns in the <code>msig$gs</code> data frame.</p>
<table class="table">
<thead><tr class="header">
<th align="left">Subset</th>
<th align="left">Description</th>
<th>Category</th>
<th>Subcategory</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Hallmark</td>
<td align="left">Curated set of gene sets</td>
<td>H</td>
<td></td>
</tr>
<tr class="even">
<td align="left">GO / BP</td>
<td align="left">Gene ontology, biological process</td>
<td>C5</td>
<td>BP</td>
</tr>
<tr class="odd">
<td align="left">GO / CC</td>
<td align="left">Gene ontology, cellular component</td>
<td>C5</td>
<td>CC</td>
</tr>
<tr class="even">
<td align="left">GO / MF</td>
<td align="left">Gene ontology, molecular function</td>
<td>C5</td>
<td>MF</td>
</tr>
<tr class="odd">
<td align="left">Biocarta</td>
<td align="left">Molecular pathways from Biocarta</td>
<td>C2</td>
<td>CP:BIOCARTA</td>
</tr>
<tr class="even">
<td align="left">KEGG</td>
<td align="left">Pathways from Kyoto Encyclopedia of Genes and Genomes</td>
<td>C2</td>
<td>CP:KEGG</td>
</tr>
<tr class="odd">
<td align="left">Reactome</td>
<td align="left">Pathways from the Reactome pathway database</td>
<td>C2</td>
<td>CP:REACTOME</td>
</tr>
</tbody>
</table>
</div>
<div class="section level4">
<h4 id="using-the-ensembl-databases-through-biomart">Using the ENSEMBL databases through biomaRt<a class="anchor" aria-label="anchor" href="#using-the-ensembl-databases-through-biomart"></a>
</h4>
<p>ENSEMBL databases for a multitude of organisms can be accessed using the R package <code>biomaRt</code>.<br>
Importantly, biomaRt allows to map different types of identifiers onto each other; this allows for example to obtain Entrez gene identifiers (required by KEGG or GO) .</p>
<p>Below, we will use biomaRt to obtain gene ontology (GO) terms and Reactome pathway IDs for genes in the <code>Egambia</code> data set, using the Entrez gene ID’s (column <code>EG</code> in the <code>Egambia</code> data set).</p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">biomaRt</span><span class="op">)</span></span>
<span><span class="va">mart</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/useMart.html" class="external-link">useMart</a></span><span class="op">(</span><span class="st">"ensembl"</span>, dataset <span class="op">=</span> <span class="st">"hsapiens_gene_ensembl"</span><span class="op">)</span></span>
<span><span class="va">bm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/getBM.html" class="external-link">getBM</a></span><span class="op">(</span>filters<span class="op">=</span><span class="st">"hgnc_symbol"</span>, </span>
<span>            values <span class="op">=</span> <span class="va">Egambia</span><span class="op">$</span><span class="va">GENE_SYMBOL</span>,</span>
<span>            attributes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span> <span class="st">"hgnc_symbol"</span>, <span class="st">"entrezgene"</span>, <span class="st">"reactome"</span>, <span class="st">"go_id"</span>, <span class="st">"name_1006"</span>, <span class="st">"go_linkage_type"</span><span class="op">)</span>,</span>
<span>            mart<span class="op">=</span><span class="va">mart</span><span class="op">)</span></span></code></pre></div>
<p>In the following code, we construct the modules data frame <code>m</code> and the gene set to gene mappings <code>m2g</code> (each twice: once for GO, and once for Reactome). We only keep the terms that have at least 10 and not more than 100 associated Entrez gene ID’s.</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m2g_r</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">bm</span><span class="op">[</span> <span class="va">bm</span><span class="op">$</span><span class="va">reactome</span> <span class="op">!=</span> <span class="st">""</span>, <span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">hgnc_symbol</span>, <span class="va">reactome</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">m2g_g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">bm</span><span class="op">[</span> <span class="va">bm</span><span class="op">$</span><span class="va">go_id</span> <span class="op">!=</span> <span class="st">""</span>, <span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">hgnc_symbol</span>, <span class="va">go_id</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ll</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lengths.html" class="external-link">lengths</a></span><span class="op">(</span><span class="va">m2g_r</span><span class="op">)</span></span>
<span><span class="va">m2g_r</span> <span class="op">&lt;-</span> <span class="va">m2g_r</span><span class="op">[</span> <span class="va">ll</span> <span class="op">&gt;=</span> <span class="fl">5</span> <span class="op">&amp;</span> <span class="va">ll</span> <span class="op">&lt;=</span> <span class="fl">250</span> <span class="op">]</span></span>
<span><span class="va">ll</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lengths.html" class="external-link">lengths</a></span><span class="op">(</span><span class="va">m2g_g</span><span class="op">)</span></span>
<span><span class="va">m2g_g</span> <span class="op">&lt;-</span> <span class="va">m2g_g</span><span class="op">[</span> <span class="va">ll</span> <span class="op">&gt;=</span> <span class="fl">5</span> <span class="op">&amp;</span> <span class="va">ll</span> <span class="op">&lt;=</span> <span class="fl">250</span> <span class="op">]</span></span>
<span></span>
<span><span class="va">m_r</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>ID<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">m2g_r</span><span class="op">)</span>, Title<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">m2g_r</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">m_g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>ID<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">m2g_g</span><span class="op">)</span>, </span>
<span>  Title<span class="op">=</span><span class="va">bm</span><span class="op">$</span><span class="va">name_1006</span><span class="op">[</span> <span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">m2g_g</span><span class="op">)</span>, <span class="va">bm</span><span class="op">$</span><span class="va">go_id</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ensemblR</span>  <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tmodGS-class.html">makeTmod</a></span><span class="op">(</span>modules<span class="op">=</span><span class="va">m_r</span>, modules2genes<span class="op">=</span><span class="va">m2g_r</span><span class="op">)</span></span>
<span><span class="va">ensemblGO</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tmodGS-class.html">makeTmod</a></span><span class="op">(</span>modules<span class="op">=</span><span class="va">m_g</span>, modules2genes<span class="op">=</span><span class="va">m2g_g</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## these objects are no longer necessary</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">bm</span>, <span class="va">m_g</span>, <span class="va">m_r</span>, <span class="va">m2g_r</span>, <span class="va">m2g_g</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="gene-ontologies-go">Gene ontologies (GO)<a class="anchor" aria-label="anchor" href="#gene-ontologies-go"></a>
</h4>
<p>GO terms are perhaps the most frequently used type of gene sets in GSE, in particular because they are available for a much larger number of organisms than other gene sets (like KEGG pathways).</p>
<p>There are many sources to obtain GO definitions. As described in the previous sections, GO’s can be also obtained from ENSEMBL via biomaRt and from MSigDB. In fact, MSigDB may be the easiest way.</p>
<p>However, GO annotations can also be obtained from AnnotationDBI Bioconductor packages as shown below. Note that the Entrez gene IDs are in the <code>EG</code> column of the <code>Egambia</code> object.</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span>
<span><span class="va">mtab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/toTable.html" class="external-link">toTable</a></span><span class="op">(</span><span class="va">org.Hs.egGO</span><span class="op">)</span></span></code></pre></div>
<p>There are over 15,000 GO terms and 250,000 genes in the <code>mtab</code> mapping; however, many of them map to either a very small or a very large number of genes. At this stage, it could also be useful to remove any genes not present in our particular data set, but that would make the resulting tmod object less flexible. However, we may be interested only in the “biological process” ontology for now.</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mtab</span> <span class="op">&lt;-</span> <span class="va">mtab</span><span class="op">[</span> <span class="va">mtab</span><span class="op">$</span><span class="va">Ontology</span> <span class="op">==</span> <span class="st">"BP"</span>, <span class="op">]</span></span>
<span><span class="va">m2g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">mtab</span><span class="op">$</span><span class="va">gene_id</span>, <span class="va">mtab</span><span class="op">$</span><span class="va">go_id</span><span class="op">)</span></span>
<span><span class="co">## remove the rather large object</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">mtab</span><span class="op">)</span> </span>
<span><span class="va">ll</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lengths.html" class="external-link">lengths</a></span><span class="op">(</span><span class="va">m2g</span><span class="op">)</span></span>
<span><span class="va">m2g</span> <span class="op">&lt;-</span> <span class="va">m2g</span><span class="op">[</span> <span class="va">ll</span> <span class="op">&gt;=</span> <span class="fl">10</span> <span class="op">&amp;</span> <span class="va">ll</span> <span class="op">&lt;=</span> <span class="fl">100</span> <span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">m2g</span><span class="op">)</span></span></code></pre></div>
<p>Using the mapping and the <code>GO.db</code> it is easy to create a module set suitable for tmod:</p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">GO.db</span><span class="op">)</span></span>
<span><span class="va">gt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/toTable.html" class="external-link">toTable</a></span><span class="op">(</span><span class="va">GOTERM</span><span class="op">)</span></span>
<span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>ID<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">m2g</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">m</span><span class="op">$</span><span class="va">Title</span> <span class="op">&lt;-</span> <span class="va">gt</span><span class="op">$</span><span class="va">Term</span><span class="op">[</span> <span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">m</span><span class="op">$</span><span class="va">ID</span>, <span class="va">gt</span><span class="op">$</span><span class="va">go_id</span><span class="op">)</span> <span class="op">]</span></span>
<span></span>
<span><span class="va">goset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tmodGS-class.html">makeTmod</a></span><span class="op">(</span>modules<span class="op">=</span><span class="va">m</span>, modules2genes<span class="op">=</span><span class="va">m2g</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">gt</span>, <span class="va">m2g</span>, <span class="va">m</span><span class="op">)</span></span></code></pre></div>
<p>This approach allows an offline mapping to GO terms, assuming the necessary DBI is installed. Using AnnotationDBI databases such as <code>org.Hs.eg.db</code> has, however, also two major disadvantages: firstly, the annotations are available for a small number of organisms. Secondly, the annotations in ENSEMBL may be more up to date.</p>
</div>
<div class="section level4">
<h4 id="kegg-pathways">KEGG pathways<a class="anchor" aria-label="anchor" href="#kegg-pathways"></a>
</h4>
<p>One way to obtain KEGG pathway gene sets is to use the MSigDB as described above. However, alternatively and for other organisms it is possible to directly obtain the pathway definitions from KEGG. The code below might take a lot of time on a slow connection.</p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">KEGGREST</span><span class="op">)</span></span>
<span><span class="va">pathways</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/KEGGREST/man/keggLink.html" class="external-link">keggLink</a></span><span class="op">(</span><span class="st">"pathway"</span>, <span class="st">"hsa"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## get pathway Names in addition to IDs</span></span>
<span><span class="va">paths</span>    <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">pathways</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">p</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/pkg/KEGGREST/man/keggGet.html" class="external-link">keggGet</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">NAME</span><span class="op">)</span></span>
<span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>ID<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">pathways</span><span class="op">)</span>, Title<span class="op">=</span><span class="va">paths</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## m2g is the mapping from modules (pathways) to genes</span></span>
<span><span class="va">m2g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">pathways</span><span class="op">)</span>, <span class="va">pathways</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## kegg object can now be used with tmod</span></span>
<span><span class="va">kegg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tmodGS-class.html">makeTmod</a></span><span class="op">(</span>modules<span class="op">=</span><span class="va">m</span>, modules2genes<span class="op">=</span><span class="va">m2g</span><span class="op">)</span></span></code></pre></div>
<p>Note that KEGG uses a slightly modified version of Entrez identifiers – each numeric identifier is preceded by a three letter species code (e.g. “hsa” for humans) followed by a colon:</p>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">eg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"hsa:"</span>, <span class="va">tt</span><span class="op">$</span><span class="va">EG</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/tmodUtest.html">tmodCERNOtest</a></span><span class="op">(</span><span class="va">eg</span>, mset<span class="op">=</span><span class="st">"kegg"</span><span class="op">)</span></span></code></pre></div>
<p>Again, the most important part is to ensure that the gene identifiers in the tmod object (<code>kegg</code> in this case) correspond to the gene identifiers in in the ordered list.</p>
</div>
<div class="section level4">
<h4 id="manual-creation-of-tmod-module-objects-msigdb">Manual creation of tmod module objects: MSigDB<a class="anchor" aria-label="anchor" href="#manual-creation-of-tmod-module-objects-msigdb"></a>
</h4>
<p>For the purposes of an example, the code below shows how to parse the XML MSigDB file using the R package XML. Essentially, this is the same code that tmodImportMsigDB is using:</p>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://www.omegahat.net/RSXML/" class="external-link">XML</a></span><span class="op">)</span></span>
<span><span class="va">foo</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/XML/man/xmlTreeParse.html" class="external-link">xmlParse</a></span><span class="op">(</span> <span class="st">"msigdb_v2022.1.Hs.xml"</span> <span class="op">)</span></span>
<span><span class="va">foo2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/XML/man/xmlToList.html" class="external-link">xmlToList</a></span><span class="op">(</span><span class="va">foo</span><span class="op">)</span></span></code></pre></div>
<p>There are over 30,000 “gene sets” (equivalent to modules in tmod) defined. Each member of foo2 is a named character vector:</p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">path1</span> <span class="op">&lt;-</span> <span class="va">foo2</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">path1</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">path1</span><span class="op">)</span></span></code></pre></div>
<p>For our example analysis, we will use only human gene sets. We further need to make sure there are no NULLs in the list.</p>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">orgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">foo2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">[</span><span class="st">"ORGANISM"</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">orgs</span><span class="op">)</span></span>
<span></span>
<span><span class="va">foo3</span> <span class="op">&lt;-</span> <span class="va">foo2</span><span class="op">[</span> <span class="va">orgs</span> <span class="op">==</span> <span class="st">"Homo sapiens"</span> <span class="op">]</span></span>
<span><span class="va">foo3</span> <span class="op">&lt;-</span> <span class="va">foo3</span><span class="op">[</span> <span class="op">!</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">foo3</span>, <span class="va">is.null</span><span class="op">)</span> <span class="op">]</span></span></code></pre></div>
<p>Next, construct the modules data frame. We will use four named fields for each vector, which contain the ID (systematic name), description, category and subcategory:</p>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">modules</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">foo3</span>, </span>
<span>  <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> </span>
<span>    <span class="va">x</span><span class="op">[</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SYSTEMATIC_NAME"</span>, <span class="st">"STANDARD_NAME"</span>, <span class="st">"CATEGORY_CODE"</span>, <span class="st">"SUBCATEGORY_CODE"</span><span class="op">)</span> <span class="op">]</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">modules</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span> <span class="st">"ID"</span>, <span class="st">"Title"</span>, <span class="st">"Category"</span>, <span class="st">"Subcategory"</span> <span class="op">)</span></span>
<span><span class="va">modules</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">modules</span>, stringsAsFactors<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">modules</span><span class="op">)</span></span></code></pre></div>
<p>Then, we create the mapping from gene sets to genes. For this, we use the <code>MEMBERS_SYMBOLIZED</code> field, which is a comma separated list of gene symbols belonging to a particular module:</p>
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m2g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">foo3</span>, </span>
<span>  <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span> <span class="va">x</span><span class="op">[</span><span class="st">"MEMBERS_SYMBOLIZED"</span><span class="op">]</span>, <span class="st">","</span> <span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">m2g</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">modules</span><span class="op">$</span><span class="va">ID</span></span>
<span></span>
<span><span class="va">mymsig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tmodGS-class.html">makeTmod</a></span><span class="op">(</span>modules<span class="op">=</span><span class="va">modules</span>, modules2genes<span class="op">=</span><span class="va">m2g</span><span class="op">)</span></span>
<span><span class="va">mymsig</span></span></code></pre></div>
<p>From now on, you can use the object mymsig with tmod enrichment tests.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="citing">Citing<a class="anchor" aria-label="anchor" href="#citing"></a>
</h2>
<p>To cite tmod, please use the following reference:</p>
<p>Zyla J, Marczyk M, Domaszewska T, Kaufmann SH, Polanska J, Weiner 3rd J. Gene set enrichment for reproducible science: comparison of CERNO and eight other algorithms. Bioinformatics. 2019 Dec 15;35(24):5146-54.</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-banchereau2012host" class="csl-entry">
Banchereau, Romain, Alejandro Jordan-Villegas, Monica Ardura, Asuncion Mejias, Nicole Baldwin, Hui Xu, Elizabeth Saye, et al. 2012. <span>“Host Immune Transcriptional Profiles Reflect the Variability in Clinical Disease Manifestations in Patients with Staphylococcus Aureus Infections.”</span> <em>PLoS One</em> 7 (4): e34390.
</div>
<div id="ref-chaussabel2008modular" class="csl-entry">
Chaussabel, Damien, Charles Quinn, Jing Shen, Pinakeen Patel, Casey Glaser, Nicole Baldwin, Dorothee Stichweh, et al. 2008. <span>“A Modular Analysis Framework for Blood Genomics Studies: Application to Systemic Lupus Erythematosus.”</span> <em>Immunity</em> 29 (1): 150–64.
</div>
<div id="ref-li2014molecular" class="csl-entry">
Li, Shuzhao, Nadine Rouphael, Sai Duraisingham, Sandra Romero-Steiner, Scott Presnell, Carl Davis, Daniel S Schmidt, et al. 2014. <span>“Molecular Signatures of Antibody Responses Derived from a Systems Biology Study of Five Human Vaccines.”</span> <em>Nature Immunology</em> 15 (2): 195–204.
</div>
<div id="ref-maertzdorf2011functional" class="csl-entry">
Maertzdorf, Jeroen, Martin Ota, Dirk Repsilber, Hans J Mollenkopf, January Weiner, Philip C Hill, and Stefan HE Kaufmann. 2011. <span>“Functional Correlations of Pathogenesis-Driven Gene Expression Signatures in Tuberculosis.”</span> <em>PloS One</em> 6 (10): e26938.
</div>
<div id="ref-subramanian2005gene" class="csl-entry">
Subramanian, Aravind, Pablo Tamayo, Vamsi K Mootha, Sayan Mukherjee, Benjamin L Ebert, Michael A Gillette, Amanda Paulovich, et al. 2005. <span>“Gene Set Enrichment Analysis: A Knowledge-Based Approach for Interpreting Genome-Wide Expression Profiles.”</span> <em>Proceedings of the National Academy of Sciences of the United States of America</em> 102 (43): 15545–50.
</div>
<div id="ref-tomfohr2005pathway" class="csl-entry">
Tomfohr, John, Jun Lu, and Thomas B Kepler. 2005. <span>“Pathway Level Analysis of Gene Expression Using Singular Value Decomposition.”</span> <em>BMC Bioinformatics</em> 6 (1): 225.
</div>
<div id="ref-tufte2014visual" class="csl-entry">
Tufte, Edward, and P Graves-Morris. 2014. <span>“The Visual Display of Quantitative Information.; 1983.”</span>
</div>
<div id="ref-weiner2016tmod" class="csl-entry">
Weiner 3rd, January, and Teresa Domaszewska. 2016. <span>“Tmod: An r Package for General and Multivariate Enrichment Analysis.”</span> <em>PeerJ Preprints</em> 2016 (09).
</div>
<div id="ref-weiner2012biomarkers" class="csl-entry">
Weiner 3rd, January, Shreemanta K Parida, Jeroen Maertzdorf, Gillian F Black, Dirk Repsilber, Anna Telaar, Robert P Mohney, et al. 2012. <span>“Biomarkers of Inflammation, Immunosuppression and Stress with Active Disease Are Revealed by Metabolomic Profiling of Tuberculosis Patients.”</span> <em>PloS One</em> 7 (7): e40221.
</div>
<div id="ref-pca3d" class="csl-entry">
Weiner, January. 2013. <em>Pca3d: Three Dimensional PCA Plots</em>.
</div>
<div id="ref-wendt1972dealing" class="csl-entry">
Wendt, Hans W. 1972. <span>“Dealing with a Common Problem in Social Science: A Simplified Rank-Biserial Coefficient of Correlation Based on the u Statistic.”</span> <em>European Journal of Social Psychology</em> 2 (4): 463–65.
</div>
<div id="ref-yamaguchi2008ifn" class="csl-entry">
Yamaguchi, Ken D, Daniel L Ruderman, Ed Croze, T Charis Wagner, Sharlene Velichko, Anthony T Reder, and Hugh Salamon. 2008. <span>“IFN-<span class="math inline">\(\beta\)</span>-Regulated Genes Show Abnormal Expression in Therapy-Na<span>ı̈</span>ve Relapsing–Remitting MS Mononuclear Cells: Gene Expression Analysis Employing All Reported Protein–Protein Interactions.”</span> <em>Journal of Neuroimmunology</em> 195 (1): 116–20.
</div>
<div id="ref-zyla2019gene" class="csl-entry">
Zyla, Joanna, Michal Marczyk, Teresa Domaszewska, Stefan HE Kaufmann, Joanna Polanska, and January Weiner 3rd. 2019. <span>“Gene Set Enrichment for Reproducible Science: Comparison of CERNO and Eight Other Algorithms.”</span> <em>Bioinformatics</em> 35 (24): 5146–54.
</div>
</div>
</div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by January Weiner.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
