<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Function reference â€¢ tmod</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><meta property="og:image" content="https://january3.github.io/tmod/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">tmod</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.50.12</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/tmod.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/user_manual/tmod_user_manual.html">tmod: A Tool for the Analysis of Gene Set Enrichments</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Function reference</h1>
    </div>

    <div class="section level2">
      <h2 id="gene-set-enrichment-tests-and-analyses">Gene set enrichment tests and analyses<a class="anchor" aria-label="anchor" href="#gene-set-enrichment-tests-and-analyses"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="tmodUtest.html">tmodUtest()</a></code> <code><a href="tmodUtest.html">tmodGeneSetTest()</a></code> <code><a href="tmodUtest.html">tmodCERNOtest()</a></code> <code><a href="tmodUtest.html">tmodPLAGEtest()</a></code> <code><a href="tmodUtest.html">tmodZtest()</a></code> <code><a href="tmodUtest.html">tmodHGtest()</a></code> 
        </dt>
        <dd>Perform a statistical test of module expression</dd>
      </dl><dl><dt>
          
          <code><a href="tmodPCA.html">tmodPCA()</a></code> 
        </dt>
        <dd>PCA plot annotated with tmod</dd>
      </dl><dl><dt>
          
          <code><a href="tmodSummary.html">tmodSummary()</a></code> 
        </dt>
        <dd>Create a summary of multiple tmod analyses</dd>
      </dl><dl><dt>
          
          <code><a href="tmodLEA.html">tmodLEA()</a></code> 
        </dt>
        <dd>Leading Edge Analysis</dd>
      </dl><dl><dt>
          
          <code><a href="tmodLEASummary.html">tmodLEASummary()</a></code> 
        </dt>
        <dd>Summary stats of a leading edge analysis</dd>
      </dl><dl><dt>
          
          <code><a href="tmodDecideTests.html">tmodDecideTests()</a></code> 
        </dt>
        <dd>Count the Up- or Down-regulated genes per module</dd>
      </dl><dl><dt>
          
          <code><a href="tmodLimmaTest.html">tmodLimmaTest()</a></code> 
        </dt>
        <dd>Run tmod enrichment tests directly on a limma object</dd>
      </dl><dl><dt>
          
          <code><a href="tmodLimmaDecideTests.html">tmodLimmaDecideTests()</a></code> 
        </dt>
        <dd>Up- and down-regulated genes in modules based on limma object</dd>
      </dl><dl><dt>
          
          <code><a href="tmodLimmaTopTable.html">tmodLimmaTopTable()</a></code> 
        </dt>
        <dd>tmod's replacement for the limma topTable function</dd>
      </dl><dl><dt>
          
          <code><a href="tmodAUC.html">tmodAUC()</a></code> 
        </dt>
        <dd>Calculate AUC</dd>
      </dl></div><div class="section level2">
      <h2 id="gene-set-module-analysis">Gene set (module) analysis<a class="anchor" aria-label="anchor" href="#gene-set-module-analysis"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="eigengene.html">eigengene()</a></code> 
        </dt>
        <dd>Calculate the eigengene of a module from a data set</dd>
      </dl><dl><dt>
          
          <code><a href="modGroups.html">modGroups()</a></code> 
        </dt>
        <dd>Find group of modules</dd>
      </dl><dl><dt>
          
          <code><a href="modOverlaps.html">modOverlaps()</a></code> 
        </dt>
        <dd>Calculate overlaps of the modules</dd>
      </dl><dl><dt>
          
          <code><a href="modcors.html">modcors()</a></code> 
        </dt>
        <dd>Module correlation</dd>
      </dl><dl><dt>
          
          <code><a href="modjaccard.html">modjaccard()</a></code> 
        </dt>
        <dd>Jaccard index for modules</dd>
      </dl></div><div class="section level2">
      <h2 id="visualizations">Visualizations<a class="anchor" aria-label="anchor" href="#visualizations"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="ggPanelplot.html">ggPanelplot()</a></code> 
        </dt>
        <dd>Create a tmod panel plot using ggplot</dd>
      </dl><dl><dt>
          
          <code><a href="ggEvidencePlot.html">ggEvidencePlot()</a></code> 
        </dt>
        <dd>Create an evidence plot for a module (ggplot2 version)</dd>
      </dl><dl><dt>
          
          <code><a href="tmodPanelPlot.html">tmodPanelPlot()</a></code> 
        </dt>
        <dd>Plot a summary of multiple tmod analyses</dd>
      </dl><dl><dt>
          
          <code><a href="pvalEffectPlot.html">pvalEffectPlot()</a></code> 
        </dt>
        <dd>Create an effect size / p-value plot</dd>
      </dl><dl><dt>
          
          <code><a href="upset.html">upset()</a></code> 
        </dt>
        <dd>Upset plot</dd>
      </dl><dl><dt>
          
          <code><a href="evidencePlot.html">evidencePlot()</a></code> 
        </dt>
        <dd>Create an evidence plot for a module</dd>
      </dl><dl><dt>
          
          <code><a href="hgEnrichmentPlot.html">hgEnrichmentPlot()</a></code> 
        </dt>
        <dd>Create a visualisation of enrichment</dd>
      </dl><dl><dt>
          
          <code><a href="modCorPlot.html">modCorPlot()</a></code> 
        </dt>
        <dd>Plot a correlation heatmap for modules</dd>
      </dl><dl><dt>
          
          <code><a href="showGene.html">showGene()</a></code> 
        </dt>
        <dd>A combined beeswarm / boxplot</dd>
      </dl><dl><dt>
          
          <code><a href="pcaplot.html">pcaplot()</a></code> 
        </dt>
        <dd>Plot a PCA object returned by prcomp</dd>
      </dl><dl><dt>
          
          <code><a href="tmodPal.html">tmodPal()</a></code> 
        </dt>
        <dd>A selection of color palettes</dd>
      </dl><dl><dt>
          
          <code><a href="tmodTagcloud.html">tmodTagcloud()</a></code> 
        </dt>
        <dd>Tag cloud based on tmod results</dd>
      </dl><dl><dt>
          
          <code><a href="simplePie.html">simplePie()</a></code> <code><a href="simplePie.html">simpleRug()</a></code> <code><a href="simplePie.html">simpleBoxpie()</a></code> 
        </dt>
        <dd>Simple Pie Chart</dd>
      </dl></div><div class="section level2">
      <h2 id="creating-and-manipulating-gene-sets-modules-">Creating and manipulating gene sets (modules)<a class="anchor" aria-label="anchor" href="#creating-and-manipulating-gene-sets-modules-"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="tmodGS-class.html">makeTmodGS()</a></code> <code><a href="tmodGS-class.html">makeTmod()</a></code> <code><a href="tmodGS-class.html">as_tmodGS()</a></code> <code><a href="tmodGS-class.html">print(<i>&lt;tmodGS&gt;</i>)</a></code> <code><a href="tmodGS-class.html">length(<i>&lt;tmodGS&gt;</i>)</a></code> <code><a href="tmodGS-class.html">`[`(<i>&lt;tmodGS&gt;</i>)</a></code> <code><a href="tmodGS-class.html">`[`(<i>&lt;tmod&gt;</i>)</a></code> 
        </dt>
        <dd>S3 class for tmod gene set collections</dd>
      </dl><dl><dt>
          
          <code><a href="makeTmodFromDataFrame.html">makeTmodFromDataFrame()</a></code> 
        </dt>
        <dd>Convert a data frame to a tmod object</dd>
      </dl><dl><dt>
          
          <code><a href="tmodImportMSigDB.html">tmodImportMSigDB()</a></code> 
        </dt>
        <dd>Import data from MSigDB</dd>
      </dl><dl><dt>
          
          <code><a href="tmod2DataFrame.html">tmod2DataFrame()</a></code> 
        </dt>
        <dd>Convert a tmod module set into a data frame</dd>
      </dl><dl><dt>
          
          <code><a href="getGenes.html">getGenes()</a></code> 
        </dt>
        <dd>Get genes belonging to a gene set</dd>
      </dl><dl><dt>
          
          <code><a href="getModuleMembers.html">getModuleMembers()</a></code> 
        </dt>
        <dd>Return the contents of a gene set</dd>
      </dl><dl><dt>
          
          <code><a href="filterGS.html">filterGS()</a></code> <code><a href="filterGS.html">showModule()</a></code> 
        </dt>
        <dd>Filter by genes belonging to a gene set from a data frame</dd>
      </dl><dl><dt>
          
          <code><a href="tmod_ids.html">tmod_ids()</a></code> <code><a href="tmod_ids.html">`tmod_ids&lt;-`()</a></code> <code><a href="tmod_ids.html">tmod_titles()</a></code> <code><a href="tmod_ids.html">`tmod_titles&lt;-`()</a></code> 
        </dt>
        <dd>Query and set IDs of gene sets in a tmodGS object</dd>
      </dl><dl><dt>
          
          <code><a href="tmod2tmodGS.html">tmod2tmodGS()</a></code> 
        </dt>
        <dd>Convert the old tmod objects to the tmodGS objects</dd>
      </dl><dl><dt>
          
          <code><a href="check_tmod_gs.html">check_tmod_gs()</a></code> 
        </dt>
        <dd>Check an object of class tmodGS</dd>
      </dl></div><div class="section level2">
      <h2 id="data-sets">Data sets<a class="anchor" aria-label="anchor" href="#data-sets"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="tmod-data.html">tmod-data</a></code> <code><a href="tmod-data.html">tmod</a></code> 
        </dt>
        <dd>Default gene expression module data</dd>
      </dl><dl><dt>
          
          <code><a href="modmetabo.html">modmetabo</a></code> <code><a href="modmetabo.html">tbmprof</a></code> 
        </dt>
        <dd>Modules for metabolic profiling</dd>
      </dl><dl><dt>
          
          <code><a href="Egambia.html">EgambiaResults</a></code> <code><a href="Egambia.html">Egambia</a></code> 
        </dt>
        <dd>Gene expression in TB patients and Healthy controls</dd>
      </dl><dl><dt>
          
          <code><a href="vaccination.html">vaccination</a></code> 
        </dt>
        <dd>Transcriptomic responses to vaccination</dd>
      </dl><dl><dt>
          
          <code><a href="cell_signatures.html">cell_signatures</a></code> 
        </dt>
        <dd>Cell type signatures</dd>
      </dl></div><div class="section level2">
      <h2 id="general-and-package-classes">General and package classes<a class="anchor" aria-label="anchor" href="#general-and-package-classes"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="tmod-package.html">tmod-package</a></code> 
        </dt>
        <dd>Transcriptional Module Analysis</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by January Weiner.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer></div>

  

  

  </body></html>

